ETCD-610: automated backups no config (#1646)
openshift/enhancements: enhancements/etcd/automated-backups-no-config.md

- As a cluster administrator I want to schedule recurring cluster backups so that I have a recent cluster state to recover from in the event of quorum loss (i.e. losing a majority of control-plane nodes).
- As a cluster administrator I want to have failure to take cluster backups for more than a configurable period to be reported to me via critical alerts.

## Proposal
- Provide a default etcd backup configuration as a `EtcdBackup` CR.
- Provide a default reliable storage mechanism to save the backup data on.
- Provide a separate feature gate for the automated backup with default config.

### Workflow Description
- The user will enable the AutomatedBackupNoConfig feature gate.
- A `EtcdBackup` CR is being installed on the cluster.
- A PVC and PV are being created to save the backup data reliably.

### API Extensions
- No API changes are required. 
- See https://github.com/openshift/enhancements/blob/ed885e10a0329a03a032b756ce9cd49a5f78af23/enhancements/etcd/automated-backups.md#L96

### Implementation Details/Notes/Constraints [optional]
- Need to agree on a default schedule.
- Need to agree on a default retention policy.

> - Several options exist for the default PVCName.

Author: @gnufied I would need your input here please

## Motivation

The [current automated backup of etcd](https://docs.openshift.com/container-platform/4.15/backup_and_restore/control_plane_backup_and_restore/backing-up-etcd.html#creating-automated-etcd-backups_backup-etcd) of an OpenShift
cluster relies on user provided configurations. 

To improve the customers experience, this work proposes taking etcd backup using default config from day one, without additional configuration from the user.

### Goals
- Backups should be taken without configuration after cluster installation
- Backups are saved to a default PersistentVolume, that could be overridden by user.
- Backups are taken according to a default schedule, that could be overridden by user.
- Backups are taken according to a default retention policy, that could be overridden by user.

### Non-Goals
- Save cluster backups to cloud storage e.g S3.
  - This could be a future enhancement or extension to the API.
- Automate cluster restoration.
- Provide automated backups for non-self hosted architectures like Hypershift.

### User Stories
- As a cluster administrator I want cluster backup to be taken without configuration.

> - As a cluster administrator I want to schedule recurring cluster backups so that I have a recent cluster state to recover from in the event of quorum loss (i.e. losing a majority of control-plane nodes).

Author: +1, I think once a day during mid-night is sufficient for most users .. I think mid-night should be a time where the cluster is not under load

> - As a cluster administrator I want to schedule recurring cluster backups so that I have a recent cluster state to recover from in the event of quorum loss (i.e. losing a majority of control-plane nodes).

Reviewer: I wouldn't assume that, but we can smear the schedule over the course of the day

## Proposal
- Provide a default etcd backup configuration as a `EtcdBackup` CR.
- Provide a default reliable storage mechanism to save the backup data on.
- Provide a separate feature gate for the automated backup with default config.

### Workflow Description
- The user will enable the AutomatedBackupNoConfig feature gate.
- A `EtcdBackup` CR is being installed on the cluster.
- A PVC and PV are being created to save the backup data reliably.

### API Extensions
- No API changes are required. 
- See https://github.com/openshift/enhancements/blob/master/enhancements/etcd/automated-backups.md#api-extensions

### Topology Considerations
TBD
#### Hypershift / Hosted Control Planes
TBD
#### Standalone Clusters
TBD
#### Single-node Deployments or MicroShift
TBD

### Implementation Details/Notes/Constraints [optional]
- Need to agree on a default schedule.
- Need to agree on a default retention policy.

- Several options exist for the default PVCName.
  - Relying on `dynamic provisioning` is sufficient, however not an option for `SNO` or `BM` clusters.

>   - Utilising `local storage operator` is a proper solution, however installing a whole operator is too much overhead.

Reviewer: why not `hostPath`  as we do it in our e2e test?
https://github.com/openshift/cluster-etcd-operator/blob/master/test/e2e/backup_test.go#L444-L487

that should be portable?

Author: so the hostPath mounts a path from the node' file system as a volume into the pod. my concerns 

 - `hostPath` could have security impact as it exposes the node's filesystem.
- there is no scheduling guarantee for the backup pod with using `hostpath` as is the case with `localVolume`.
- Using `localVolume`, the node affinity with the PV will force the backup pod to be scheduled on a specific node, where the volume is attached.
- `localVolume` allows using a separate disk as PV, unlike `hostPath` which mounts a folder from the node's FS.
- `localVolume` is handled by the PV controller and the scheduler in different manner, in fact it was created to resolve issues with `hostPath`

That being said, if we were to use `localVolume`, we need to find a solution for balancing the backups across the master nodes, some ideas could be 

- create a PV for each master node using `localVolume` and the backup controller from CEO should take care of balancing the backups across the `available` and `healthy` volumes.
- the controller should keep the most recent backup on a healthy node available for restoration.
- the controller should skip an unhealthy master node from taking a backup

Reviewer: can you list your concerns and this solution as an alternative here in the doc as well, please?

Author: added these comments to the `alternatives` and `Risks` sections 👍🏽

    - Cons :-
      - If the master node where the PV is mounted became unhealthy/unavailable/unreachable etc.
      - The backups are no longer accessible, also taking new backups is no longer possible.
      - Ideally, the backups should be taken on a round-robin fashion on each master node to avoid overwhelming a specific etcd member more than others.
      - Also spreading the backups among all etcd cluster members provide guarantees for disaster recovery in case of losing two members at the same time.
      - However, using the `local volume` option over all the master nodes will be complicated and error prune.

### Risks and Mitigations

When the backups are configured to be saved to a `local` type PV, the backups are all saved to a single master node where the PV is provisioned on the local disk.

In the event of the node becoming inaccessible or unscheduled, the recurring backups would not be scheduled. The periodic backup config would have to be recreated or updated with a different PVC that allows for a new PV to be provisioned on a node that is healthy.

> If we were to use `localVolume`, we need to find a solution for balancing the backups across the master nodes, some ideas could be

Reviewer: how would we restore from a localVolume? Is the data stored somewhere in `/var/`?

Author: so the localVolume is being mounted anywhere as specified in the Pod. In fact, you can mount a whole external drive. 

iiuc, `hostPath` supports only mount points from the node's FS

Reviewer: imagine we have to run the restore script, how would we get the local volume mounted into a recovery machine? would we create another static pod that would mount the localVolume?

I imagine the local storage plugin must be running along with the kubelet for that to work? @jsafrane / @gnufied maybe you guys can elaborate a bit how that works, I don't have much time this week to read through the CSI implementation :racehorse:

Reviewer 2: pardon my ignorance for not being familiar with etcd deployment, but even if there is a default SC available, why are we backing up just one replica of etcd? I thought, we would want to backup all replicas no?  Such as - what if in case of network partition or something, the replica we are backing up is behind?

So IMO it sounds whether we are using LocalVolume or something else, we should always be creating backups across master nodes? Is that accurate?

Reviewer 2: You can also use a directory as a `localvolume` if statically provisioning a local PV. So basically `path: ` field points to a directory. The difference with `hostPath` volumes is, the container using local-volume doesn't have to run with elevated privileges.

Reviewer: > I thought, we would want to backup all replicas - no? Such as - what if in case of network partition or something, the replica we are backing up is behind?

Snapshots are always generated from the leader, so it's never behind. We only want to _store_ the snapshot on all nodes. 

>  Local Volumes are a inbuilt feature of k8s and hence no additional plugin is necessary

that's great to hear, thanks for the docs :) 

> the container using local-volume doesn't have to run with elevated privileges.

the restore script/ pod will have to run with elevated permissions anyway, because of static pods and host networking. We don't have that permission constraint at all, which is why hostPaths are somewhat convenient.

Author: @tjungblu 

> imagine we have to run the restore script, how would we get the local volume mounted into a recovery machine? would we create another static pod that would mount the localVolume?

So the localVolume will be created with `Retain` policy. This way the `PV` will still have the same files and can be mounted in the new etcd pod and the backup will exist. 

However, if the whole machine is unhealthy/unreachable, then the backup will not be accessible.

Reviewer: > ... and can be mounted in the new etcd pod and the backup will exist.

check the restore script, you will need to unpack the snapshot before you can run any pod

Author: @gnufied in order to make sure the backups are accessible, and that we can round-robin the backups across all master nodes, which are up-to-date with the `etcd-cluster` leader

Is it possible to 

- Create StatefulSet across the master nodes, where the PVC template uses the `localVolume` to provision PV.
- Since we need to do the backup according to a schedule as per [EtcdBackupSpec](https://github.com/openshift/api/blob/0689f006bcdeccf685c75438835beffa8bb14db7/config/v1alpha1/types_backup.go#L46), the issue to manage the backup pods within the STS using the CronJob
- I am not aware if this is possible, but at least, we can generate `Event` from the CronJob and the STS could react to these event by taking a backup

Author: > check the restore script, you will need to unpack the snapshot before you can run any pod

We can use `init-container` to unpack the backup, and then the etcd-pod can start ? 

Otherwise, we can start a fresh etcd-pod and then run `etcdctl restore` ? 

Please correct me if I am wrong

Reviewer 3: > * Create StatefulSet across the master nodes, where the PVC template uses the `localVolume` to provision PV.

I believe you mean DaemonSet across the master nodes :wink:

Author: I actually thought about this as well, but there is no `PersistentVolumeClaim` template on DaemonSet.

Having a separate PVC & PV for each master node would be ideal for spreading the backups across the master nodes. 

That being said, according to the latest update, we are going to use `SideCar` container && `hostPath` volume

> - create a PV for each master node using `localVolume` and the backup controller from CEO should take care of balancing the backups across the `available` and `healthy` volumes.

Reviewer: would a DaemonSet do that for us?

Author: Good point. 

- We can use DS and use node affinity to run the DS pods only on master nodes. However, I am not aware of how the storage will be handled in this case. 

- Is there is a way to keep the backups in round-robin fashion among all master nodes ? 

- Also what about STS ? wdyt ?

Reviewer: yep, STS could also work, especially with the storage. Would also be useful to have this compared to the static pod/hostpath approach

Author: Alright, I think if we have STS, then the Pods, PVC, and PVs are being managed together

I think we can in this case create backups on RoundRobin manner across the master nodes. 

For BM, I think we can still use local volume as storage option and being managed by STS

Reviewer 2: How does this round robin backup work?

Reviewer 2: I read the original enhancement - https://github.com/openshift/enhancements/blob/master/enhancements/etcd/automated-backups.md and I understand this now better. But still this begs the question - how do you know which snapshot to restore from? What if you take snapshot of a etcd replica which was behind and is just catching up with other master nodes? Is that possible?

Author: So we will make sure that the backup is taken from the a member whose log is identical to the `leader`. This way we make sure that we are not lagging behind.

But, could this approach work ?

Reviewer 3: Why do you want to handle the balancing? That's not what you should care for at all.

Author: Yes I agree, therefore the `SideCar` approach is the most appropriate

- the controller should keep the most recent backup on a healthy node available for restoration.
- the controller should skip an unhealthy master node from taking a backup

### Drawbacks

## Design Details

> - Add default values to current `Etcdbackup Spec` using annotation.

Author: @openshift/openshift-team-etcd I added this section after discussion on the `Architecture Call`

Input is from @soltysh @jsafrane 

- It is recommended to use `default` values on current API.
- Do not rely on CVO on managing the default config.
- Use storage solution based on OCP  :- 
  - Dynamic provisioning on Cloud based.
  - Local Volume for SNO / BM.

Reviewer: let's say we have to restore from a snapshot, how can we access the local volume?

Reviewer 3: The general flow should be like follows:
1. All the EtcdBackup fields are defaulted, where possible. That's why I'm proposing to put this functionality behind already existing `AutomatedEtcdBackup` feature gate, rather than introducing a new one. 
2. The `pvcName` will be the one that will have to by populated automatically by the CEO, unless a users sets one. It will use:
  - the available default storage class in the cluster (storage team can provide you with where you should be looking at it);
  - if above is not available, fallback to using a localVolume, as explained by Hemant above;
  - add a warning condition to the CEO (at least for that localVolume), reporting that we're working off of default etcd backup, and that the cluster admin should verify correctness of that configuration.

- Maintain the reconciliation logic with Cluster-etcd-operator.
 

`Storage` solution used is based on two different designs, based on the OCP variant used 

#### Cloud based OCP

##### Pros

> - Utilising `Dynamic provisioning` is best option on cloud based OCP.

Reviewer: how is the customer able to access a dynamically created PV to get their snapshots for restoration?

Author: Gr8 question, so I have defaulted the `RetentionPolicy` in my sample configs to `Retain`. 

This way the PV contents will never be erased automatically. 

Now to answer the restoration 

- If the cluster still running, the PV can be attached to any node, as long as it is in the same availability zone.
- If the cluster is completely down, the PV can be accessed by the CU, it will never be delete unless manually.

Reviewer: it's not about the retention, it's about the access and mounting possibilities 

> If the cluster still running, the PV can be attached to any node, as long as it is in the same availability zone.

that's quite a bad constraint for a disaster recovery procedure, isn't it? :)

Author: :D :D  .. I know, well the best option imho is to push backups to **Remote storage** option, then even in case of the whole cluster is down, a new installation can be restored using the remote backup. 

However, in my discussion yesterday, the remote storage was not an option for BM. Also they recommended to keep it simple not too complicated 

If it were to me, I would create a side car which pushes the backups to remote storage, I think this option would work for any OCP variant. The master nodes dont have to be in the cloud, the side car could authenticate and push backups regardless of OCP underlying infrastructure, wdyt

Reviewer 2: If dyanmically provisioned volumes are available in the cluster, it is 100% likely that, cluster has Remote Storage and hence backups are available even if cluster is down.

Author: I actually want to fall back to the `localVolume` approach, since having two separate approaches is harder to maintain. 

I believe the localVolume should be sufficient. However, if we could utilize STS across all the master nodes would be ideal in situation where we lose one or more master nodes.

Reviewer 2: If I were to design this, I would definitely use two different backup strategies. It is far more reliable to use dynamically provisioned PVCs for backups. They can be snapshotted and can be accessible if node goes down. 

It seems to me that - any hostpath/localvolume based approach is basically inferior. If last backup was taken on leader and then leader node goes down, then so does our backup(other nodes may have slightly behind backups). So hostpath/localvolume requires fundamentally different recovery mechanism. 

IMO - if we take other components that use persistent storage in Openshift, if customers don't provide Storage then no persistent configuration is created. Prometheus or image-registry, they all require storage. And they don't do anything automatically.

So, what I am saying is - we should probably limit scope of this KEP only to environments where an StorageClass is available. If no StorageClass is configured then no automatic backups are taken. 

We should not take upon ourselves to decide a storage strategy for customer. cc @tjungblu

Reviewer 2: So to summarize - I would simply not bother to configure backups via localvolume/hostpath in environments where no storage is available and solve that problem via documentation or ask customer to configure storage. I do not think, we should automatically configuring local-storage for these clusters.

Author: I actually agree that using dynamic provisioning and CSI snapshot is vital for the automated backup. 

I also agree to limit the `automated backup` for clusters with `dynamic provisioning` only. 

wdyt @tjungblu @hasbro17 @dusk125 @soltysh

On the other hand, the backups may **not** be up-to-date since the snapshot might be lagging behind the `WAL`. Therefore, it is recommended to use this approach alongside the Automated Backup enabled using the `EtcdBackup` CR.
Since this work will be enabled with no configuration, it is possible to use define a default values for the `Scheule`, `Retention` independently.

Since cluster backups must be stored on a persistent storage, and each storage solution has pros and cons, several approaches that have been investigated. See [Alternatives](#Alternatives)

This work utilizes `hostPath` approach, since the `SideCar` container has access to the etcd pod's file system.

### Open Questions

This enhancement adds on previous work [automated backup of etcd](https://docs.openshift.com/container-platform/4.15/backup_and_restore/control_plane_backup_and_restore/backing-up-etcd.html#creating-automated-etcd-backups_backup-etcd).
Therefore, it is vital to use the same feature gate for both approaches. The following issues need to be clarified before implementation

> * How to distinguish between `NoConfig` backups and configs that are triggered using `EtcdBackup` CR.

Author: @soltysh Would you kindly help me answering these questions ?

Reviewer: we will call it name="default" and in the controller we'll skip the reconciliation of that CRD

Author: so if the sidecar is enabled we skip the backup using the controller, or we skip the default backups if the controllers is enabled ?

- Backups are taken according to a default schedule, that could be overridden by user.
- Backups are taken according to a default retention policy, that could be overridden by user.

### Non-Goals
- Save cluster backups to cloud storage e.g S3.
  - This could be a future enhancement or extension to the API.
- Automate cluster restoration.
- Provide automated backups for non-self hosted architectures like Hypershift.

### User Stories
- As a cluster administrator I want cluster backup to be taken without configuration.
- As a cluster administrator I want to schedule recurring cluster backups so that I have a recent cluster state to recover from in the event of quorum loss (i.e. losing a majority of control-plane nodes).
- As a cluster administrator I want to have failure to take cluster backups for more than a configurable period to be reported to me via critical alerts.

## Proposal
- Provide a default etcd backup configuration as a `EtcdBackup` CR.
- Provide a default reliable storage mechanism to save the backup data on.
- Provide a separate feature gate for the automated backup with default config (under discussion).

### Workflow Description

> - The user will enable the AutomatedBackupNoConfig feature gate (under discussion).

Reviewer 3: Given that `AutomatedEtcdBackup` is still a tech preview feature, I'd consider expanding that functionality with the default configuration, rather than adding another one. This will save a lot of problems for some of the fields in the API you're planning to provide default values for.

Author: +1, I will fix this 👍🏽

- A `EtcdBackup` CR is being installed on the cluster.
- In a default config scenario, the use should not add any config.
- A PVC and PV are being created to save the backup data reliably.

### API Extensions
- No API changes are required. 
- See https://github.com/openshift/enhancements/blob/master/enhancements/etcd/automated-backups.md#api-extensions

### Topology Considerations
TBD
#### Hypershift / Hosted Control Planes
TBD
#### Standalone Clusters
TBD
#### Single-node Deployments or MicroShift

> TBD

Reviewer 3: I believe the strategy will be different for different topology, so I'd expect an outline how each topology will be handled wrt defaults.

Author: So we are planning to use **SideCar Container** within each etcd pod manifest, and the backup will be save to a `hostPath` volume. 

iiuc in this approach the strategy will be the same regardless of the topology, or ?

Reviewer 3: Yes, it should, just make sure to call it out explicitly.

Reviewer 3: If you're taking the exact same approach across all supported topologies, I'd at minimum add here links to the design details, pointing out that the solution is orthogonal to topology of the cluster.

Reviewer 4: Probably not applicable to Microshift - it doesn't use CEO so it can't schedule backups iiuc. Lets mention that in the enhancement.

SNO probably won't be significantly different from HA cluster

Author: Yes it is the same approach, that why we took the sidecar approach

Reviewer 5: This comment appears to be unresolved.

Since this work will be enabled with no configuration, it is possible to use define a default values for the `Scheule`, `Retention` independently.

Since cluster backups must be stored on a persistent storage, and each storage solution has pros and cons, several approaches that have been investigated. See [Alternatives](#Alternatives)

This work utilizes `hostPath` approach, since the `SideCar` container has access to the etcd pod's file system.

### Open Questions

This enhancement adds on previous work [automated backup of etcd](https://docs.openshift.com/container-platform/4.15/backup_and_restore/control_plane_backup_and_restore/backing-up-etcd.html#creating-automated-etcd-backups_backup-etcd).
Therefore, it is vital to use the same feature gate for both approaches. The following issues need to be clarified before implementation

* How to distinguish between `NoConfig` backups and configs that are triggered using `EtcdBackup` CR.
* Upon enabling the `AutomatedBackup` feature gate, which approach should be used and according to what criteria.
* How should the current backup controller distinguish between both approaches.

> * How to disable the `NoConfig` behaviour, if a user want to.

Reviewer: that depends on how we install that "default" CRD, but it would either be a deletion of that CRD (so we need to ensure we don't install it back again) or an annotation that will disable it

Author: i will remove this since we do not install CR at all, it is a side car container that exist in all etcd pod manifests

Author: since there is no default CRD and we use the sidecar approach, we could 

- re-deploy the etcd deployment without the `backupnoconfig` sidecar upon `automatedbackup` CRD installation. 
- as you suggested, we can add annotation that disables the sidecar 

I vote for the second approach, to avoid rolling out a new revision

> * As the backups are being taken in the `NoConfig` approach among all master nodes, how to detect the most up-to-date backup.

Reviewer: since all nodes on schedule will take almost the same backup (+- a few dozen revisions) I don't really see why that would be a problem?

Author: +1, removing ...

If we were to use `localVolume`, we need to find a solution for balancing the backups across the master nodes, some ideas could be

- create a PV for each master node using `localVolume` and the backup controller from CEO should take care of balancing the backups across the `available` and `healthy` volumes.
- the controller should keep the most recent backup on a healthy node available for restoration.
- the controller should skip an unhealthy master node from taking a backup

### Drawbacks

## Design Details

- Add default values to current `Etcdbackup Spec` using annotation.
- Maintain the reconciliation logic with Cluster-etcd-operator.

### Initial Storage Design proposal
- Several options exist for the default `PVCName`.
  - Relying on `dynamic provisioning` is sufficient, however not an option for `SNO` or `BM` clusters.
  - Utilising `local storage operator` is a proper solution, however installing a whole operator is too much overhead.

>   - The most viable solution to cover all OCP variants is to use `local volume`.

Reviewer 2: You don't have to necessarily install local-storage-operator to use/consume local-storage. For simple one-off use cases like this, it may be possible that etcd-backup operator or something creates static-pv as documented - https://docs.openshift.com/container-platform/4.15/storage/persistent_storage/persistent_storage_local/persistent-storage-local.html#local-create-cr-manual_persistent-storage-local and uses it to perform etcd backups.

#### Single-node and Bare metal OCP
    Only possible option is to use `local volume`. 

### Open Questions
How to use remote storage for backups on `BM` and `SNO`. 

## Test Plan

An e2e test will be added to practice the scenario as follows 

- Enable AutomatedBackupNoConfig FeatureGate.
- Verify that a `Etcdbackup` CR has been installed.
- Verify that a PV, PVC has been created.
- Verify that a backup has been taken.
- Verify that the backups are valid according to the retention policy.

## Graduation Criteria
TBD

### Dev Preview -> Tech Preview
TBD

### Tech Preview -> GA
TBD

### Removing a deprecated feature
TBD

## Upgrade / Downgrade Strategy
TBD

## Version Skew Strategy
TBD

## Operational Aspects of API Extensions
TBD

#### Failure Modes
TBD

## Support Procedures
TBD

## Implementation History
TBD

## Alternatives
### hostPath

Utilising `hostPath` volumes as a storage solution for the backups.

`hostPath` mounts a path from the node's file system as a volume into the pod.

It supports all OCP variants, including `SNO` and `BM`.  

> However, I am strongly against using it for the following reasons

Reviewer 3: ```suggestion
However, the following reasons suggest against that solution:
```

Author: this whole has been removed, radical changes :D :D

      - `hostPath` could have security impact as it exposes the node's filesystem.
      - No scheduling guarantees for the pod using `hostpath` as is the case with `localVolume`. The pod could be scheduled on a different node from where the hostPath volume exist.
      - On the other hand, using `localVolume` and the node affinity within the PV manifest forces the backup pod to be scheduled on a specific node, where the volume is attached.
      - `localVolume` allows using a separate disk as PV, unlike `hostPath` which mounts a folder from the node's FS.

>       - `localVolume` is handled by the PV controller and the scheduler in different manner, in fact it was created to resolve issues with `hostPath`

Reviewer 3: You forgot to add the locality of the backup. Iow. if it happens that backup is kept in the current leader, and that dies, we're loosing that data with it.

Author: So the current approach will take backups among all master nodes. Actually we have another issue now, how to distinguish which backup is most up-to-date since we have backups from all master nodes :)

reviewers:
  - "@soltysh"
  - "@dusk125"
  - "@hasbro17"
  - "@tjungblu"
  - "@williamcaban"
approvers:
  - "@soltysh"
  - "@dusk125"
  - "@hasbro17"
  - "@tjungblu"
  - "@williamcaban"
api-approvers:
  - "@soltysh"
creation-date: 2024-06-17
last-updated: 2024-06-17
tracking-link:
  - https://issues.redhat.com/browse/ETCD-609
see-also:
  - "https://issues.redhat.com/browse/OCPSTRAT-1411"
  - "https://issues.redhat.com/browse/OCPSTRAT-529"
---

# Automated Backups of etcd with No Config

## Summary

To enable automated etcd backups of an Openshift cluster from day one after installation, 
without provided configuration from the user.

## Motivation

The current [automated backup of etcd](https://docs.openshift.com/container-platform/4.15/backup_and_restore/control_plane_backup_and_restore/backing-up-etcd.html#creating-automated-etcd-backups_backup-etcd) of an OpenShift
cluster relies on user provided configurations. 

To improve the customers experience, this work proposes taking etcd backup using default config from day one, without additional configuration from the user.

### Goals

* Backups should be taken without configuration after cluster installation from day 1.

> * Backups are saved to a default PersistentVolume, that could be overridden by user.

Author: so I am checking the `Backup` CR name and the `periodicbackup` controller ignores the CR with name `default`

This way we have both approaches alongside each other, and they both work independently

* Backups are taken according to a default schedule, that could be overridden by user.
* Backups are taken according to a default retention policy, that could be overridden by user.

### Non-Goals

* Save cluster backups to remote cloud storage (e.g. S3 Bucket).

>   - This could be a future enhancement or extension to the API.

Reviewer 3: As discussed in one of our calls, the ability to guarantee a persistent storage for any supported cluster installation is very small. So I wouldn't even call out future extension. It's the cluster admin role to ensure that the PVCname is backed by a solid storage.

  - Having backups in an independent storage from OCP cluster is crucial for Disaster Recovery scenarios.
* Automate cluster restoration.
* Provide automated backups for non-self hosted architectures like Hypershift.

## Proposal

To enable automated etcd backups of an Openshift cluster, a default etcd backup option to be introduced. 
Using a `SideCar` container within each etcd pod, backups could be taken with no config from user.

### User Stories

* As a cluster administrator I would like OCP cluster backup to be taken without configuration.
* As a cluster administrator I would like to schedule recurring OCP cluster backups so that I have a recent cluster state to recover from in the event of quorum loss (i.e. losing a majority of control-plane nodes).
* As a cluster administrator I would like to have alerts upon failure to take OCP cluster backup.

### API Extensions

> No [API](https://github.com/openshift/enhancements/blob/master/enhancements/etcd/automated-backups.md#api-extensions) changes are required, since this approach work independently with default config.

Reviewer 3: I believe you'll need to introduce the defaults in the API, no?

Author: So i will create a default CR in CEO .. i will not use kubebuilder markers on the API 

```
apiVersion: config.openshift.io/v1alpha1
kind: Backup
metadata:
  name: default
  annotations:
    default: "true"
spec:
  etcd:
    schedule: "20 4 * * *"
    timeZone: "UTC"
    retentionPolicy:
      retentionType: RetentionNumber
      retentionNumber:
        maxNumberOfBackups: 5
```

Reviewer 6: When you create this default CR, is this going to be in all clusters? Are we happy doing this for existing clusters?

Would it be better for the installer to create this object so that it's only created on new clusters? Con: We can't manage it to change it on existing clusters (though we probably don't want to)

### Topology Considerations
TBD
#### Hypershift / Hosted Control Planes
TBD
#### Standalone Clusters
TBD
#### Single-node Deployments or MicroShift
TBD
### Workflow Description
TBD

## Design Details

A `SideCar` container within each etcd pod in Openshift cluster is being added in order to take backups without any provided configuration from user's perspective.

In order to alleviate the overhead on etcd server, the side container takes the backup by simply copying the snapshot file.
This approach is being used by `microshift`, and it has minimal overhead as it is a simple file system copy operation.
Since the `SideCar` is being deployed alongside each etcd cluster member, it is possible to keep backups across all master nodes.

On the other hand, the backups may **not** be up-to-date since the snapshot might be lagging behind the `WAL`. Therefore, it is recommended to use this approach alongside the Automated Backup enabled using the `EtcdBackup` CR.

> Since this work will be enabled with no configuration, it is possible to use define a default values for the `Scheule`, `Retention` independently.

Reviewer 3: It looks like this is completely different from what I remember we talked about last time. I'd like you to change the initial part of this document to clearly explain this mechanism is orthogonal to the user-configured backups.

Since cluster backups must be stored on a persistent storage, and each storage solution has pros and cons, several approaches that have been investigated. See [Alternatives](#Alternatives)

This work utilizes `hostPath` approach, since the `SideCar` container has access to the etcd pod's file system.

### Open Questions

> This enhancement adds on previous work [automated backup of etcd](https://docs.openshift.com/container-platform/4.15/backup_and_restore/control_plane_backup_and_restore/backing-up-etcd.html#creating-automated-etcd-backups_backup-etcd).

Reviewer 3: Based on the above statements they are two separate mechanisms, are they not?

without provided configuration from the user.

## Motivation

The current [automated backup of etcd](https://docs.openshift.com/container-platform/4.15/backup_and_restore/control_plane_backup_and_restore/backing-up-etcd.html#creating-automated-etcd-backups_backup-etcd) of an OpenShift
cluster relies on user provided configurations. 

To improve the customer's experience, this work proposes taking etcd backup using default config from day one, without additional configuration from the user.

### Goals

* Backups should be taken without configuration after cluster installation from day 1.
* Backups are saved to a default location on master's node disk using `hostPath` volume.
* Backups are taken according to a default schedule.
* Backups are taken according to a default retention policy.

### Non-Goals

* Save cluster backups to remote cloud storage (e.g. S3 Bucket).
  - This could be a future enhancement or extension to the API.
  - Having backups in an independent storage from OCP cluster is crucial for Disaster Recovery scenarios.
* Automate cluster restoration.
* Provide automated backups for non-self hosted architectures like Hypershift.

## Proposal

> To enable automated etcd backups of an Openshift cluster, a default backup option is to be taking sing a `SideCar` container within each etcd pod, with no configuration from user.

Reviewer 6: Does this document describe how the existing backups are taken? Or is there something that can be read to understand how that works today?

### Topology Considerations
TBD
#### Hypershift / Hosted Control Planes
TBD
#### Standalone Clusters
TBD
#### Single-node Deployments or MicroShift
TBD
### Workflow Description
TBD

## Design Details

A `SideCar` container is being added within each etcd pod in Openshift cluster in order to take backups without any provided configuration from user's perspective.

As the `SideCar` is being deployed alongside each etcd cluster member, it is possible to keep backups across all master nodes.

Since this work will be enabled with no configuration, it is possible to use define a default values for the `Schedule`, `Retention` independently.

This work utilizes `hostPath` approach, since the `SideCar` container has access to the etcd pod's file system.

Before deciding upon this approach, several approaches that have been investigated. See [Alternatives](#Alternatives).

### Open Questions

* How to distinguish between `NoConfig` backups and configs that are triggered using `EtcdBackup` CR.
  * The `NoConfig` backups relies on `EtcdBackup` CR with `default` name.

>   * The cluster-etcd-operator reacts to the `default` CR by deploying the backup sidecar containers alongside each etcd member. 

Reviewer 6: Does it need to be a sidecar? Is it going to run always? Could it not be a job? (I assume the existing backups happen via a job?)

* Upon enabling the `AutomatedBackup` feature gate, which approach should be used and according to what criteria.

>   * As the `Noconfig` backups is orthogonal to the [automated backup of etcd](https://docs.openshift.com/container-platform/4.15/backup_and_restore/control_plane_backup_and_restore/backing-up-etcd.html#creating-automated-etcd-backups_backup-etcd), it has been decided to use the same feature gate.

Reviewer 6: That means you have to get both features ready and tested before either can be promoted. Strongly advise to use a separate gate

Author: it is one feature gate, that has been used in https://github.com/openshift/enhancements/blob/master/enhancements/etcd/automated-backups.md 

The `PeriodicBackupController` is feature gated with it https://github.com/openshift/cluster-etcd-operator/blob/ddbab6c9a4e65d3dbcd5a312caabd5e05879ef95/pkg/operator/periodicbackupcontroller/periodicbackupcontroller.go#L111

Reviewer 6: Are you and your PM/lead willing to promote the periodic backups only once this feature has also passed full QE testing?

Author: I am not the one who answers this question to be honest :D :D

Reviewer 6: Who is?

Author: the PM or Lead :D :D

Reviewer 6: And who are they? 😂

Author: @hasbro17 @tjungblu @racedo

  

* How should the current backup controller distinguish between both approaches.
  * The `NoConfig` backups relies on `EtcdBackup` CR with `default` name.
  

* How to disable the `NoConfig` behaviour, if a user want to.

>   * This has not been decided or implemented yet. 

Reviewer 6: If the installer created the object, and it was unmanaged, the user would just delete it

### Implementation Details/Notes/Constraints

The `NoConfig` backups builds atop of the existing [automated backup of etcd](https://docs.openshift.com/container-platform/4.15/backup_and_restore/control_plane_backup_and_restore/backing-up-etcd.html#creating-automated-etcd-backups_backup-etcd).

> The `PeriodicBackupController` within `cluster-etcd-operator` checks for `EtcdBackup` CR with `default` name. Once found, a sidecar container is deployed within each etcd pod in Openshift cluster.

Reviewer 6: This suggests the existing backups also use a sidecar?

The container runs a backup server with the default schedule as defined in default `EtcdBackup` CR. 

The backups are being stored on the local disk of each master node independently, using `hostPath` volume. This replication is required for disaster recovery scenarios where more than one of the master nodes has been lost.

> A default `EtcdBackup` CR is used for the `NoConfig` backups. A default schedule of daily at midnight has been chosen. See below

Reviewer 6: Midnight UTC could be the middle of peak operational hours for a cluster deployed in APAC or NASA. Is it advisable to take a backup when you're at the peak?

Are backups disruptive to reads/writes at all?

You'll likely need a way for this to be configured or changed

```yaml
apiVersion: config.openshift.io/v1alpha1
kind: Backup
metadata:
  name: default
spec:
  etcd:
    schedule: "0 0 * * *"
    timeZone: "UTC"
```

### Risks and Mitigations

## Test Plan

An e2e test will be added to practice the scenario as follows 

- Enable AutomatedBackupNoConfig FeatureGate.
- Verify that a `Etcdbackup` CR has been installed.
- Verify that a backup has been taken.
- Verify that the backups are valid according to the retention policy.

## Graduation Criteria
TBD

### Dev Preview -> Tech Preview
TBD

### Tech Preview -> GA
TBD

### Removing a deprecated feature
TBD

## Upgrade / Downgrade Strategy
TBD

## Version Skew Strategy
TBD

## Operational Aspects of API Extensions
TBD

#### Failure Modes
TBD

## Support Procedures
TBD

## Implementation History
TBD

### Drawbacks

## Alternatives

**Ideally**, the storage solution to use would have the following characteristics

> - Backups should be taken on a round-robin fashion on each master node to avoid overwhelming a specific etcd member more than others.

Reviewer 6: Have we considered the storage space required for the backups, is there configuration for retention? What is the impact of adding this data to the nodes?

- One fits all storage solution, to be used across all Openshift variants including Bare metal and Single Node Openshift.

Below is each storage solution, with its pros and cons.

### Dynamic provisioning using CSI

As Openshift cluster that are based on cloud infrastructure offers dynamic storage provisioning capabilities based on CSI, allocating persistent storage for the etcd backups is trivial.

Depending on the storage capabilities offered by the cloud provider, the allocated persistent volume could be attached to any of the master nodes and the backups can be used by any etcd member on any of the master nodes for restoration.

* Pros
  - Utilising CSI storage is ideal as it is well tested and allocating storage required defining a `PVC` manifest only.
  - Some storage solutions based on CSI allows a `PV` to be attached to multiple nodes. This is ideal for both taking backup and restoration from various master nodes, in case one of the nodes is unreachable.
  - It allows recovery even on disaster scenarios, where no master node is accessible, since the storage is allocated on the provider side.
* Cons
  - CSI is not an option for OCP `BM` or `SNO`.

>   - Some cloud providers allocate storage to be accessible only within the same availability zone. Since Openshift cluster allocate each master node on a different zone for high availability purposes, the backup will not be accessible if the master node where the backup was taken is no longer accessible.

Reviewer 6: This can depend on the kind of storage, I believe on AWS EBS is AZ tied, but EFS is not, you could also in theory leverage S3 which also isn't AZ tied. Other clouds have similar concepts IIRC

Author: @deads2k wdyt about this ? 

The remote storage is the best option for restoring a cluster from a disaster scenario 

Also utilizing remote storage eliminates the need for replicating the backup across all master nodes

Author: this is the approach taken by `hypershift` https://github.com/openshift/hypershift/blob/main/etcd-backup/etcdbackup.go

### Local Storage

> Relying on local storage works for all Openshift variants (i.e. cloud based, SNO and BM). There are three possible approaches using local storage, each is detailed below.

Reviewer 6: On SNO, this doesn't seem particularly helpful. What is the recommendation going to be for SNO users?

Reviewer 4: SNO users can create a PV with `hostPath` and a PVC pointing to it. Also, most SNOs have LVM-backed provisioner so its not a major problem

Author: so `hostPath` works well for it 👍🏽

Had the Pod been replaced, the new Pod is guaranteed to be scheduled on the same node, where the backup exists.
This deterministic scheduling behaviour is a property of `localVolume`, which is not the case with `hostPath` volumes.
`localVolume` is handled by the PV controller and the scheduler in different manner, in fact it was created to resolve issues with `hostPath`

* Pros
  - It supports all Openshift variants, including `SNO` and `BM`.
  - The Backup Pod will be scheduled always to this master node where the backup data exists.
  - Using a separate disk for backups (e.g. SSD) is possible, since local volume allows mounting it into a Pod.
  - Having the deterministic Pod scheduling is vital, since the restoration needs the recent backup in place to work.
* Cons
  - If the master node where the `PV` is mounted became unhealthy/unavailable/unreachable, then the backups are no longer accessible, also taking new backups is no longer possible.

#### Host Path volumes

A `hostPath` volume mounts a path from the node's file system as a volume into a Pod.

* Pros
  - It supports all Openshift variants, including `SNO` and `BM`.
* Cons

>   - `hostPath` could have security impact as it exposes the node's filesystem.

Reviewer 6: What about the other way around, are backups encrypted? What happens if someone gets hold of a backup by plucking it from a hostpath?

Reviewer: if you get hostpath access, you might as well just take the etcd data dir directly. Backups are never encrypted, but the secrets within _should_ be.

>   - No scheduling guarantees for the pod using `hostpath` as is the case with `localVolume`. The pod could be scheduled on a different node from where the hostPath volume exist.

Reviewer 6: If you have a particular want, you could schedule the pods using the CEO to guarantee where they end up

Author: yes , this has been already implemented

### Hybrid approach

As shown above, dynamic provisioning is the best storage solution to be used, but it is not viable among all Openshift variants.
However, a hybrid solution is possible, in which dynamic provisioning is being used in cloud based Openshift, while utilising local storage for BM and SNO.

> Relying on `infrastructure` resource type, we can create the storage programmatically according to the underlying infrastructure and Openshift variant.

Reviewer 6: So you plan to create cloud based PVs on the clouds?

Author: not at all, just hostPath :D

### StatefulSet approach

A `StatefulSet` could be deployed among all master nodes, where each backup pod has its own `PV`. This approach has the pros of spreading the backups among all master nodes.

> The complexity come from the fact that the backups are being triggered by a `CronJob` which spawn a `Job` to take the actual backup, by deploying a Pod.

Reviewer 6: CEO could add the nodeName to the podSpec in the Job when it creates it and handle this

Author: lets continue discussing this on the test result https://github.com/openshift/enhancements/pull/1646#issuecomment-2564491734 

resolving this ...

### Non-Goals

* Save cluster backups to remote cloud storage (e.g. S3 Bucket).
  - This could be a future enhancement or extension to the API.
  - Having backups in an independent storage from OCP cluster is crucial for Disaster Recovery scenarios.
* Automate cluster restoration.
* Provide automated backups for non-self hosted architectures like Hypershift.

## Proposal

To enable automated etcd backups of an Openshift cluster from Day 1, without user supplied configuration, this work proposes a default etcd backup option.

The default backup option utilizes a sidecar container within each etcd pod. The container is responsible for maintaining backups within a `hostPath` volume on each master node's disk.

This approach guarantees have backups across all master nodes which is vital in disaster recovery scenarios.

### User Stories

* As a cluster administrator, I would like OCP cluster backup to be taken without configuration from day one.
* As a cluster administrator, I would like to schedule recurring OCP cluster backups so that I have a recent cluster state to recover from in the event of quorum loss (i.e. losing a majority of control-plane nodes).

### API Extensions

> No [API](https://github.com/openshift/enhancements/blob/master/enhancements/etcd/automated-backups.md#api-extensions) changes are required.

Reviewer 6: How would a user configure the destination of the backups without an API? If they wanted to move them to persistent storage? Looks like we are leveraging an existing API, so perhaps it's worth mentioning that this is re-using the existing API at this stage, rather than saying there is no API?

Reviewer 7: >Looks like we are leveraging an existing API, so perhaps it's worth mentioning that this is re-using the existing API at this stage, rather than saying there is no API?

@Elbehery I think this is currently a source of confusion in this enhancement. Either we're using the existing `config.openshift.io/v1alpha1` `Backup` API or not? This seems to be saying it bothways.
>To distinguish default no-config backups from user configured counterparts, a config.openshift.io/v1alpha1 Backup CR's name must be default

Also relevant to my other comment is that this also brings into question whether or not this feature should have it's own feature gate `AutomatedBackupNoConfig ` when you're saying this is reliant on the API behind the existing `AutomatedEtcdBackup` feature gate.

https://github.com/openshift/enhancements/pull/1646/files#r1781963569

Reviewer: > How would a user configure the destination of the backups without an API? 

I thought the goal of this epic was to have **no configuration**, why would that need an API?

Author: @JoelSpeed so we are using the same API introduced in `automated-backups` work. 

> re-using the existing API at this stage, rather than saying there is no API?

+1, I am changing this now. 

So what I meant that we did not introduce any API changes from `automated-backups`

Author: @hasbro17 

- Yes, we are using the same API `config.openshift.io/v1alpha1` `Backup`, I will clarify this now all over the text 👍🏽 
-  So as discussed before, we agreed to use the same feature gate, that's why we are utilizing the `PeriodicBackupController`. Also having two separate feature gates, imho;  will cause problems in the future as this is the same work.

### Topology Considerations
TBD
#### Hypershift / Hosted Control Planes
TBD
#### Standalone Clusters
TBD
#### Single-node Deployments or MicroShift
TBD
### Workflow Description
TBD

## Design Details

A sidecar container is being deployed within each etcd pod in Openshift cluster in order to take backups according to default configuration using `config.openshift.io/v1alpha1 Backup`.

To distinguish default _no-config_ backups from user configured counterparts, a `config.openshift.io/v1alpha1 Backup` CR's name must be `default`.

> Initially, the sidecar container within etcd pods is not running (i.e. disabled). However, creating a `Backup` CR with `name=default` enables the container.

Reviewer 6: Why not use the existing mechanisms for backing up which don't rely on sidecars, rather than inventing a new pattern? Is that documented somewhere?

Author: > How large is an average OpenShift etcd database, and then from that, how large is a backup expected to be?

it is within range [8,32] GB. 

the backup is way smaller than this, I think `perf&scale` can give us some exact numbers, if needed 

also the `Pruning` functionality makes sure we never run out of space

Author: > Why not use the existing mechanisms for backing up which don't rely on sidecars, rather than inventing a new pattern? Is that documented somewhere?

- so the existing mechanism will be also available, this work provides the ability to have `AutomatedBackup` with no configuration from CUs

- Also this allows to have `AutomatedBackup` from Day1, after cluster installation. 

- You can find in `Alternatives` section the result of more than a month of investigation. We settled on this in the end, since it is the most straightforward  and allows to have backups on all master nodes, also it works for `BM` and `SNO` clusters

Also disabling the container from the etcd pod upon error is not recommended, unless requested by the user. As discussed above, enabling and disabling the sidecar container leads to deploying a new version of the static pods.
This scenario leads to a duration of cluster unavailability. 

As shown above, it is necessary to report the no-config backup sidecar failure without failing the container. Below are possible resolution scenarios. 

- Log the error upon failure
  - A log parser could be used in order to detect the error, without failing the container.
  
- Update Cluster-etcd-Operator StatusConditions
  - Reporting the condition and degrading the operator could be a solution. 

- Use a Prometheus gauge in order to report the container situation. 

### Implementation Details/Notes/Constraints

The _no-config_ backups builds atop of the existing [automated backup of etcd](https://docs.openshift.com/container-platform/4.15/backup_and_restore/control_plane_backup_and_restore/backing-up-etcd.html#creating-automated-etcd-backups_backup-etcd).

> The `PeriodicBackupController` within **cluster-etcd-operator** checks for `Backup` CR with `name=default` within every sync. Upon applying a `Backup` CR with `name=default`, the _no-config_ backup sidecar container is being enabled within each etcd pod in Openshift cluster.

Reviewer 6: This just seems like a special case within the periodic backups no? Why do we need a new mechanism rather than the periodic backup controller just handling running backups on a default schedule?

Author: good question 👍🏽  

So, the `default` CR is the same type of CR (i.e. config.openshift.io/v1alpha1 Backup).  The reasons to use `PeriodicBackupController` are 

- Having two separate controllers to monitor the same CR (config.openshift.io/v1alpha1 Backup) type is `anti-pattern` 
- If the user Creates or Deletes the default no-config backup, it should not influence the Cron based Backups, as they are orthogonal.
- Having the same controller manages the all the Backup CRs is ideal for `FeatureGate` reasons.

Reviewer 6: I appreciate the idea to use the same controller, that makes sense, and I think my question is agreeing with what you've said.

My point here was that, this part of the EP describes a mechanism for initiating backups. I'm trying to work out how this compares to the mechanism in use for the periodic backups.

As in, how are the periodic backups implemented, do they use a set of static pods?

And then from that, if there is a difference in how the backup is implemented for default vs periodic, why do we need that difference?

Author: > As in, how are the periodic backups implemented, do they use a set of static pods?

No, they are not. So the high-level design is being depicted [here](https://docs.google.com/document/d/1iIyJYAl2G70RPsPY3wF65i06PGFWBJs-Hk7z68tFPEM/edit?usp=sharing) within the `Troubleshooting` section

So `PeriodicBackups` are being implemented using two levels 

- A `OneTime Backup` triggers a Job, which triggers a Pod which runs `backup.sh` script.
- `PeriodicBackups` is a `CronJob` which executes the `OneTime Backup` upon each trigger.

This implementation does not rely on StaticPods at all. 

> And then from that, if there is a difference in how the backup is implemented for default vs periodic, why do we need that difference?

So the main motivations behind `default` method are

- Provide Backups from Day1 without user supplied configurations
- Having the Backups replicated, for disaster recovery scenarios 

iiuc, OCP variants such as `SNO` and `BM` can not use PVC with external storage, the only options are `hostPath` and `localVolume`. 

For cloud based OCP, having a PVC bound to external storage eliminates the need for replicating the backups over control-plane nodes, since the PV is on cloud provider's side, and backup are accessible on Disaster scenarios. 

On the other hand, using `hostPath` on the `default` method is a way to provide backup replicas, while having the same method applicable on `SNO` and `BM`

Reviewer 4: Why do we need `default` instance at all? I imagine we can have multiple instances with different schedules and retention policies (i.e. weekly and monthly snapshots)

Author: I actually think these ideas worth exploring 

cc @tjungblu

The `Backup` configurations such as `scheudle`, `timezone` and `retention policy`, are being supplied to the _no-config_ container as flags. 
According to user's preference, the default configuration could be modified by updating the CR.

> The backups are being stored on the local disk of each master node independently, using `hostPath` volume. This replication is required for disaster recovery scenarios where more than one of the master nodes has been lost.

Reviewer 6: Haven't had any answers to these questions yet

> The backups are being stored on the local disk of each master node independently, using `hostPath` volume. This replication is required for disaster recovery scenarios where more than one of the master nodes has been lost.

Author: > Haven't had any answers to these questions yet

Deep apologies, I am replying to all the Qs now :)

> The backups are being stored on the local disk of each master node independently, using `hostPath` volume. This replication is required for disaster recovery scenarios where more than one of the master nodes has been lost.

Author: > Does this pose risk in terms of credentials that the backup job would need?

So no credentials are being exposed at all, not even the path of the backup

> The backups are being stored on the local disk of each master node independently, using `hostPath` volume. This replication is required for disaster recovery scenarios where more than one of the master nodes has been lost.

Author: > Tying the backup pod to the main etcd pod seems to be particularly fraught with danger, in terms of enabling and disabling, and when the backup has issues. It would make more sense for it to be separate no?

Actually the new approach that uses DaemonSet has shown very solid deployment. The  pods had only `1` restart over a period of 45 minutes, and the reason of the restart on all the pods (i.e. a Backup Pod is being deployed on every control-plane node), was caused by `node` 

see https://github.com/openshift/cluster-etcd-operator/pull/1354#issuecomment-2408661261

> The backups are being stored on the local disk of each master node independently, using `hostPath` volume. This replication is required for disaster recovery scenarios where more than one of the master nodes has been lost.

Reviewer 6: I was looking specifically for answers to https://github.com/openshift/enhancements/pull/1646#discussion_r1781269175 which I still don't think are answered

> The backups are being stored on the local disk of each master node independently, using `hostPath` volume. This replication is required for disaster recovery scenarios where more than one of the master nodes has been lost.

Author: Deep apologies again, I thought I commented on this 

> How frequent are the default backups? What is the impact to the etcd cluster when the backup occurs, does it freeze the DB for any period?

So the default configurations are not decided yet, I have this in the `open questions` section, we still need to have consensus over the default configurations 

> What happens when the control plane is resized using CPMS, is there any risk of a period where there are no backups left because CPMS got rid of them?

iiuc, CPMS deletes the master nodes upon resizing and replace them with newly created once. In this case, and since the backups are stored on `hostPath` volume, which is just a folder on the node's file-system, the backups will be lost :/

> The backups are being stored on the local disk of each master node independently, using `hostPath` volume. This replication is required for disaster recovery scenarios where more than one of the master nodes has been lost.

Reviewer 4: >What is the impact to the etcd cluster when the backup occurs

this may be quite important - if the IO hit significant we may want to restrict the frequency of backups (i.e. at least daily)

> The backups are being stored on the local disk of each master node independently, using `hostPath` volume. This replication is required for disaster recovery scenarios where more than one of the master nodes has been lost.

Author: This needs figures from Perf&Scale imo

A `Backup` CR with default configuration below.

```yaml
apiVersion: config.openshift.io/v1alpha1
kind: Backup
metadata:
  name: default
spec:
  etcd:
    schedule: "*/5 * * * *"
    timeZone: "UTC"
    retentionPolicy:
      retentionType: RetentionNumber
      retentionNumber:
        maxNumberOfBackups: 3
```

### Risks and Mitigations

## Test Plan

An e2e test will be added to practice the scenario as follows 

> - Enable AutomatedBackupNoConfig FeatureGate.

Reviewer 7: This is the first mention of this feature gate I'm seeing in this proposal. I don't see any details on this new feature gate and why we are or not using the existing one for automated backups.

Did I miss how you resolved these discussions on using the existing vs a new feature gate?
https://github.com/openshift/enhancements/pull/1646#discussion_r1654846128
https://github.com/openshift/enhancements/pull/1646#discussion_r1711358625
https://github.com/openshift/enhancements/pull/1646#discussion_r1654864573

Author: > This is the first mention of this feature gate I'm seeing in this proposal. I don't see any details on this new feature gate and why we are or not using the existing one for automated backups.

Sorry, I have fixed the type, it is the same FeatureGate introduced in https://github.com/openshift/enhancements/blob/master/enhancements/etcd/automated-backups.md

Author: > Did I miss how you resolved these discussions on using the existing vs a new feature gate?

I have replied now to https://github.com/openshift/enhancements/pull/1646#discussion_r1711358625

The other two threads have been discussed on slack before, the outcome was to use the same feature gate, to avoid issues when going GA and since it is the same work, using a separate feature gate does not make sense

## Proposal

To enable automated etcd backups of an Openshift cluster from Day 1, without user supplied configuration, this work proposes a default etcd backup option.

The default backup option utilizes a sidecar container within each etcd pod. The container is responsible for maintaining backups within a `hostPath` volume on each master node's disk.

This approach guarantees have backups across all master nodes which is vital in disaster recovery scenarios.

### User Stories

* As a cluster administrator, I would like OCP cluster backup to be taken without configuration from day one.
* As a cluster administrator, I would like to schedule recurring OCP cluster backups so that I have a recent cluster state to recover from in the event of quorum loss (i.e. losing a majority of control-plane nodes).

### API Extensions

No [API](https://github.com/openshift/enhancements/blob/master/enhancements/etcd/automated-backups.md#api-extensions) changes are required.

### Topology Considerations
TBD
#### Hypershift / Hosted Control Planes
TBD
#### Standalone Clusters
TBD
#### Single-node Deployments or MicroShift
TBD
### Workflow Description
TBD

> ## Design Details

Reviewer 4: Can we predict the size of the backup and ensure we won't run out of space when storing a backup?

Author: well, I don't think we will run out of space for number of backups, 

however, there is a backup-pruning functionality that trims the backups to avoid keeping them infinitely .. this way we won't run out of space

Reviewer 6: How large is an average OpenShift etcd database, and then from that, how large is a backup expected to be?

Author: iirc, OCP starts initially with 8GB backend-quota, we have limits up to 32GB of the size of db

Author: That's good point, so given the current DB size, I can estimate beforehand, given the retention policy, the size of the DB, and the disk space on the node, I can estimate beforehand 

I think it worth an alert ,or ? 

cc @tjungblu

Author: created https://issues.redhat.com/browse/ETCD-692 to track this 👍🏽

Reviewer 5: alerting on disk space is good. Please add to GA criteria.

Author: I created https://issues.redhat.com/browse/ETCD-692 to track it, how to label it with GA ? 🤔

Reviewer 8: > ...how to label it with GA ?

Add text describing this as a GA requirement down in the:

> ```markdown
> ### Tech Preview -> GA
> ```

Section of this enhancement, where you currently just say `TBD`.

Author: done 👍🏽

> ## Design Details

Reviewer 4: Does backup file has a way to ensure file correctness? Perhaps sha256sum file generated automatically?

Author: we can add this, computing the SHA not a big issue 

cc @openshift/openshift-team-etcd  wdyt

Author: I actually like this idea, shall we add it to our GA backlog ? 

cc @openshift/openshift-team-etcd wdyt

Reviewer 5: This is a very good point. A hash should be added to the design details.

Author: ok, so after taking the backup, I compute the hash and write it on disk ?

Reviewer 8: to disk, and also possibly to Backup `status` (which is currently empty):

```console
api$ git grep -A1 'type BackupStatus struct' config/v1alpha1
config/v1alpha1/types_backup.go:type BackupStatus struct {
config/v1alpha1/types_backup.go-}
```

just to get the hash stored somewhere further away from the backup file being hashed?

Author: +1 @tjungblu, I think this is not needed 👍🏽

A sidecar container is being deployed within each etcd pod in Openshift cluster in order to take backups according to default configuration using `config.openshift.io/v1alpha1 Backup`.

To distinguish default _no-config_ backups from user configured counterparts, a `config.openshift.io/v1alpha1 Backup` CR's name must be `default`.

Initially, the sidecar container within etcd pods is not running (i.e. disabled). However, creating a `Backup` CR with `name=default` enables the container.

Once enabled, the backups are being taken on each master node according to the provided `schedule` within the CR. Moreover, the backups are being maintained according to the supplied retention policy. 

The default CR's configuration could be overridden by the user according to their preference.

> The backups are being stored within a `hostPath` volume. The volume is a directory within each master node's file system.

Reviewer 4: How are files expected to be named?

Author: ```
oc rsh -c etcd-backup-server -n openshift-etcd pod/etcd-ip-10-0-109-64.ec2.internal 
sh-5.1# ls -l /var/backup/etcd/
total 0
drwxr-xr-x. 2 root root 96 Aug 26 15:10 2024-08-26_151000
drwxr-xr-x. 2 root root 96 Aug 26 15:15 2024-08-26_151500
drwxr-xr-x. 2 root root 96 Aug 26 15:20 2024-08-26_152000
```

see the comments on this [PR](https://github.com/openshift/cluster-etcd-operator/pull/1325#issuecomment-2310546816), it has a test for the whole work

Author: it follows this naming https://github.com/openshift/cluster-etcd-operator/blob/0b7f488071c585a131483ab4e3dae050faeed32c/test/e2e/backup_test.go#L671-L675

Reviewer 4: ah, okay, so it has a date which is useful. I think its worth adding the IP too, so that when admins collect backups from different nodes they wouldn't have them mixed. Not sure if this is important though (the DB should be consistent across other nodes too, so the backups are probably interchangeable)

Author: so the backups are taken by each backup pod from the local etcd member on the same node 

 I can add the ip-address, it is not an issue, 

Wdyt @hasbro17 @tjungblu

Before deciding upon this approach, several approaches that have been investigated. See [Alternatives](#Alternatives).

### Open Questions

> Currently, the no-config backup sidecar container exits upon error. Since the container is being deployed within the etcd Pod, the whole Pod will never be `Running` as long as the container is failing.

Author: cc @hasbro17 @tjungblu @dusk125 @vrutkovs @gnufied @jsafrane @simonpasquier  @dgrisonnet @wking 

Would appreciate your input

Reviewer 4: Probably the container should never fail - and instead report the error via CEO status

Author: shall we go degraded as well ?

Author: @JoelSpeed ptal :)

> Currently, the no-config backup sidecar container exits upon error. Since the container is being deployed within the etcd Pod, the whole Pod will never be `Running` as long as the container is failing.

Reviewer: I see two options:
* As Vadim mentioned, the sidecar will never fail and report any issues into a status of some sort
* We create an entirely new static pod yaml that only contains the backup container

Reviewer 6: > We create an entirely new static pod yaml that only contains the backup container

Does it even need to be a static pod? But yes, I think I already encouraged having a separate deployment or otherwise rather than having this as a sidecar, so +1 from me here

Author: @tjungblu iiuc, it should be a staticPod in order to have access to the node's file-systems, so that we access `etc-data` dir .. correct ? 

btw, I also think reporting failure through CEO's `statusCondition` is fair enough. This follows the same approach taken by `controllers` upon failure

Reviewer 6: You could access the data dir without it being a static pod, it could be a deployment, a job, or anything that generates a pod really

> btw, I also think reporting failure through CEO's statusCondition is fair enough. This follows the same approach taken by controllers upon failure

Does this pose risk in terms of credentials that the backup job would need?

Author: I think we can mount a `hostPath` Volume on the etcd backup directory. We can access it from backup-pod running on all master nodes 

so we can 

- use deployment
- node-affinity so that a backup pod is being scheduled on every node where etcd is running 
- another option is to use `PodGroup`, I am not sure it can manage mirrored pod which corresponds to  staticPods.

> Does this pose risk in terms of credentials that the backup job would need?

I don't think mounting `hostPath` volume would need any credentials, please correct me if I'm  wrong

Once the container fails, it will be automatically restarted according to exponential backoff strategy. This scenario leads to `CrashLoopBackOff` and the etcd Pod will never be `Running`. 

As the no-config backup sidecar is being deployed alongside each etcd member, consequently an error within this container leads to unavailable etcd cluster. Hence, failing the container upon error is not a viable option.
Also disabling the container from the etcd pod upon error is not recommended, unless requested by the user. As discussed above, enabling and disabling the sidecar container leads to deploying a new version of the static pods.

> This scenario leads to a duration of cluster unavailability. 

Reviewer: not sure where you got that notion from?

Author: so during redeploying staticPods upon _enabling_ or _disabling_ **etcd-backup-container** , the cluster is `api-server` is unavailable  for the revision rollout duration

I have seen this consistently  during testing, pleaser correct me if I am wrong

Reviewer 6: Tying the backup pod to the main etcd pod seems to be particularly fraught with danger, in terms of enabling and disabling, and when the backup has issues. It would make more sense for it to be separate no?

As shown above, it is necessary to report the no-config backup sidecar failure without failing the container. Below are possible resolution scenarios. 

- Log the error upon failure
  - A log parser could be used in order to detect the error, without failing the container.
  
- Update Cluster-etcd-Operator StatusConditions
  - Reporting the condition and degrading the operator could be a solution. 

> - Use a Prometheus gauge in order to report the container situation. 

Reviewer: how can prometheus scrape a crash looping container metric?

Author: so if the container is down over a duration threshold, then it is considered failed 

btw, I am brainstorming  :D

The _no-config_ backups builds atop of the existing [automated backup of etcd](https://docs.openshift.com/container-platform/4.15/backup_and_restore/control_plane_backup_and_restore/backing-up-etcd.html#creating-automated-etcd-backups_backup-etcd).

The `PeriodicBackupController` within **cluster-etcd-operator** checks for `Backup` CR with `name=default` within every sync. Upon applying a `Backup` CR with `name=default`, the _no-config_ backup sidecar container is being enabled within each etcd pod in Openshift cluster.
The `Backup` configurations such as `scheudle`, `timezone` and `retention policy`, are being supplied to the _no-config_ container as flags. 
According to user's preference, the default configuration could be modified by updating the CR.

The backups are being stored on the local disk of each master node independently, using `hostPath` volume. This replication is required for disaster recovery scenarios where more than one of the master nodes has been lost.

A `Backup` CR with default configuration below.

```yaml
apiVersion: config.openshift.io/v1alpha1
kind: Backup
metadata:
  name: default
spec:
  etcd:
    schedule: "*/5 * * * *"

>     timeZone: "UTC"

Reviewer 4: Do we need a separate timezone setting? Where is it being used?

Author: this is the default, if this setting has been changed, the parser we use should be taking care of this iiuc 

we use github.com/robfig/cron/v3

Reviewer: this originated from the fact that we transformed those into k8s cronjobs, which have those field:
https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/#time-zones

Reviewer 4: I don't see any benefit for the user to run backup in EST when the rest of the cluster is in UTC for instance - it would make configuration less confusing imo

Author: so u mean to default to the cluster's timezone ?

    retentionPolicy:
      retentionType: RetentionNumber
      retentionNumber:
        maxNumberOfBackups: 3
```

### Risks and Mitigations

## Test Plan

An e2e test will be added to practice the scenario as follows 

- Enable AutomatedBackupNoConfig FeatureGate.
- Verify that a `Backup` CR has been installed.
- Verify that backups have been taken successfully.
- Verify that backups are being maintained according to the retention policy.

> 

Reviewer 4: e2e test should also attempt to restore etcd from backup to make sure backup is not corrupted

Author: the restore is out of scope so far iiuc 

cc @openshift/openshift-team-etcd wdyt

Reviewer: we have separate restore tests that take a fresh backup and restore from it

Reviewer 5: > e2e test should also attempt to restore etcd from backup to make sure backup is not corrupted

> we have separate restore tests that take a fresh backup and restore from it

We need a check to be sure our backup is valid.  If that can be done without a restore, ok. If not, then wiring up a restore after the backup is taken would work really well.

Author: we can instead 

- compare with the backup on other nodes
- compare with the data on the current node

iiuc, the restoration  should be done on a fresh node. If the test will allocate a new master node, the test time will be significantly  long, wdyt !!

Author: we could utilize the `verifyBackup` cmd from CEO [here](https://github.com/openshift/cluster-etcd-operator/blob/e706f3f63235faecd1c2c51b418d78f1f2d6cfb4/pkg/cmd/verify/backupstorage.go#L52)

Actually, it is being used in the CronBased backup Job [here](https://github.com/openshift/cluster-etcd-operator/blob/ae0e36ce8beae851b31975169da72134be18a3eb/bindata/etcd/cluster-backup-job.yaml#L13)

`localVolume` is handled by the PV controller and the scheduler in different manner, in fact it was created to resolve issues with `hostPath`

* Pros
  - It supports all Openshift variants, including `SNO` and `BM`.
  - The Backup Pod will be scheduled always to this master node where the backup data exists.
  - Using a separate disk for backups (e.g. SSD) is possible, since local volume allows mounting it into a Pod.
  - Having the deterministic Pod scheduling is vital, since the restoration needs the recent backup in place to work.
* Cons
  - If the master node where the `PV` is mounted became unhealthy/unavailable/unreachable, then the backups are no longer accessible, also taking new backups is no longer possible.

#### Host Path volumes

A `hostPath` volume mounts a path from the node's file system as a volume into a Pod.

* Pros
  - It supports all Openshift variants, including `SNO` and `BM`.
* Cons
  - `hostPath` could have security impact as it exposes the node's filesystem.
  - No scheduling guarantees for the pod using `hostpath` as is the case with `localVolume`. The pod could be scheduled on a different node from where the hostPath volume exist.

### Hybrid approach

> As shown above, dynamic provisioning is the best storage solution to be used, but it is not viable among all Openshift variants.

Reviewer: that sentence makes little sense, given we settled on using hostpaths

Author: +1, updating ....

Author: done 👍🏽

---
title: automated-backups-no-config
authors:
  - "@elbehery"
reviewers:
  - "@soltysh"
  - "@dusk125"
  - "@hasbro17"
  - "@tjungblu"
  - "@williamcaban"
approvers:
  - "@soltysh"
  - "@dusk125"
  - "@hasbro17"
  - "@tjungblu"

>   - "@williamcaban"

Author: done 👍🏽

api-approvers:

>   - "@soltysh"

Author: you are reviewer everywhere :)

creation-date: 2024-06-17

> last-updated: 2024-06-17

Author: done 👍🏽

- Spreading the backups among all etcd cluster members provide guarantees for disaster recovery in case of losing two members at the same time.
- One fits all storage solution, to be used across all Openshift variants including Bare metal and Single Node Openshift.

According to these requirement, only `hostPath` and `localVolume` storage could be utilized. 
However, if the master node where the `PV` is mounted became unhealthy/unavailable/unreachable, then the backups are no longer accessible, also taking new backups is no longer possible.
Hence, `hostPath` has been used as it is the only viable option that could be utilized on all Openshift variants. 

### Workload

In order to take backups for Etcd data store and cluster resources, a workload must run on the cluster while having access to Etcd and cluster resources.
Since Etcd data store of an Openshift cluster runs on `Pod` similar to other workload, several constructs such as `Deployment`, `DaemonSet`, `StatefulSet` and `SideCar` could be used.

This section discusses two approaches (i.e. `SideCar` and `DaemonSet`), which are implemented and tested, explaining the Pros and Cons of each method.  

> Before deciding upon these approaches, several approaches have been investigated. See [Alternatives](#Alternatives).

Author: > Are there concerns about having this disk space available? Do we need to make sure there is overhead on control plane nodes? 32GB is not trivially available by default I would wager

Added to open questions. However, the pruning functionality has an option to prune the backups according to size

> Before deciding upon these approaches, several approaches have been investigated. See [Alternatives](#Alternatives).

Author: > Before merging, one of these should be moved to the alternatives section

+1, so @hasbro17 wanted to open the discussion about which approach to use before deciding the final one, that's why I have added both of them here with their Pros&Cons 

I personally vote for the `DaemonSet` approach, since deploying the side car within etcd static pod involves rolling out new revision upon enabling / disabling, and also imposing a risk of bringing the whole etcd cluster down upon any error from the backup container

**Pros**

Etcd pod is a static pod, and the availability of the etcd cluster is vital for cluster stability. As the etcd cluster is being managed by `Cluster-etcd-operator`, and a `Deployment` resource manages the _mirrored_ etcd pods, any changes in the pod manifest results into rolling out a new revision, which leads to a duration of etcd-cluster unavailability. Consequently, the `API-Server` can not process any write requests, and read requests are being handled; if any, from the `API-Server` cache.   

The backup sidecar container is being `disabled` by default. Upon creating a `Backup` CR on an Openshift cluster, the `PeriodicBackupController` reacts by enabling the container. This leads to a change in etcd static pod manifest, and a new revision is being rolled out. As explained above, such action causes cluster instability and `API-Server` unavailability.

Additionally, any error within the backup sidecar, the container goes into `CrashLoopBackOff`, which leads to the whole `Pod` reports `NotReady` status. Consequently, the whole `Etcd-cluster` can not server requests, as well as `API-Server`.

#### DaemonSet manages an Etcd-backup pod within every control-plane node.

> This approach has been initiated during reviews of the `Sidecar` method. A reviewer has recommended to separate the backup container from the static pod to mitigate the issues shown above.

Author: yes I omitted it .. I was working till very late, sorry was inappropriate  :)

By setting the `nodeSelector` to the master node label (i.e. `node-role.kubernetes.io/master`), guarantees that an Etcd-backup pod will be scheduled on every control-plane node.

**Pros**

This method has been implemented and tested. The same container used in the Sidecar approach has been utilised. However, the scheduling is being managed by a DaemonSet resource.
Upon creating a `Backup` CR on an Openshift cluster, the `PeriodicBackupController` reacts by creating a DaemonSet resource, with the `etcd-backup-server` container and the configurations applied in the CR.
During [testing](https://github.com/openshift/cluster-etcd-operator/pull/1354#issuecomment-2408661261), this method has shown a solid and stable deployment, where the Backup pods shown only `1` restart over a duration of `45` minutes. Moreover, the cause of the restart on all the Backup pods were due to `node` issue, not the backup container.

> **Cons**

Author: yes, they are all being taken simultaneously

> **Cons**

Reviewer 6: And, when a backup occurs, does that prevent reads and/or writes to the etcd instance/cluster?

> **Cons**

Author: gr8 Question. 

So I don't think so, citing https://etcd.io/docs/v3.5/op-guide/recovery/

```
Note that taking the snapshot from the member/snap/db file might lose data that has not been written yet, but is included in the wal (write-ahead-log) folder.
```

Read/Write request has to go through Raft and the WAL, since the backup relies on the snapshot, the backup might be lacking the records that have been replicated through the WAL, but not applied to the StateMachine yet (i.e. bbolt) 

cc @hasbro17 @dusk125 @tjungblu

### Open Questions

### Implementation Details/Notes/Constraints

The _no-config_ backups builds atop of the existing [automated backup of etcd](https://docs.openshift.com/container-platform/4.15/backup_and_restore/control_plane_backup_and_restore/backing-up-etcd.html#creating-automated-etcd-backups_backup-etcd).

The `PeriodicBackupController` within **cluster-etcd-operator** checks for `Backup` CR with `name=default` within every sync. Upon applying a `Backup` CR with `name=default`, the _no-config_ backup sidecar container is being enabled within each etcd pod in Openshift cluster.
The `Backup` configurations such as `scheudle`, `timezone` and `retention policy`, are being supplied to the _no-config_ container as flags. 
According to user's preference, the default configuration could be modified by updating the CR.

The backups are being stored on the local disk of each master node independently, using `hostPath` volume. This replication is required for disaster recovery scenarios where more than one of the master nodes has been lost.

A `Backup` CR with default configuration below.

```yaml
apiVersion: config.openshift.io/v1alpha1
kind: Backup
metadata:
  name: default

> spec:

Reviewer 6: How many of these fields are defaulted vs required, is something creating this default for the user?

Author: > How frequent are the default backups? What is the impact to the etcd cluster when the backup occurs, does it freeze the DB for any period?

So the default values are not being decided yet, and  I added this to the OpenQuestion section for discussion 

> What happens when the control plane is resized using CPMS, is there any risk of a period where there are no backups left because CPMS got rid of them?

Actually the backups will be gone, since they are being saved on `hostPath` which is on the node's file system 

Since CPMSO will delete the master nodes, the backup will be gone 

However, since this works orthogonal to the `CronJob` implementation, any backups that are saved on the mounted PVC will be available .. Also a continuation of this work is to use `external target` for saving backups, for `Disaster Recovery` scenarios, where all the control-plane nodes are gone

- One fits all storage solution, to be used across all Openshift variants including Bare metal and Single Node Openshift.

According to these requirement, only `hostPath` and `localVolume` storage could be utilized. 
However, for `localVolume` storage, if the master node where the `PV` is mounted became unhealthy/unavailable/unreachable, then the backups are no longer accessible, also taking new backups is no longer possible.
Hence, `hostPath` has been used as it is the only viable option that could be utilized on all Openshift variants. 

### Workload

In order to take backups for Etcd data store and cluster resources, a workload must run on the cluster while having access to Etcd and cluster resources.
Since Etcd data store of an Openshift cluster runs on `Pod` similar to other workload, several constructs such as `Deployment`, `DaemonSet`, `StatefulSet` and `SideCar` could be used.

This section discusses two approaches (i.e. `SideCar` and `DaemonSet`), which are implemented and tested, explaining the Pros and Cons of each method.  
Before deciding upon these two methods, several other approaches have been investigated. See [Alternatives](#Alternatives).

> #### SideCar container within all Etcd-static-pod

Author: I would appreciate your final opinion, which approach should be used and merged ? 

I personally vote for the `DaemonSet` approach, for the following 

- It mitigates  the issues of the `StaticPod` approach, such as rolling out a new revision upon enabling / disabling the `etcd-backup-server`

Reviewer 4: I like daemonset approach. With 2 node OpenShift we can make sure advisor node is excluded - this approach is more flexible than sidecar

Author: I also think the same, I was astonished by how stable it is. 

I think it worth testing this during Cluster-Disruption 

So we make noise on the control-plane nodes, and see how this affects the Backup pods, wdyt ? 

I think if the Etcd cluster is under load, we should shut down the backup, because this will overload the cluster, which is already suffering

Reviewer 5: Move this to alternatives, let's go with daemonset

Author: @vrutkovs  @deads2k would you kindly vote [here](https://github.com/openshift/enhancements/pull/1646#issuecomment-2567130894)

During [testing](https://github.com/openshift/cluster-etcd-operator/pull/1354#issuecomment-2408661261), this method has shown a solid and stable deployment, where the Backup pods shown only `1` restart over a duration of `45` minutes. Moreover, the cause of the restart on all the Backup pods were due to `node` issue, not the backup container.

_**Pros**_

As the Backup pods are being scheduled independently of the etcd pods, any error within the Backup pods does not influence cluster Openshift cluster availability.

Moreover, as the Backup pods are being scheduled within every master nodes, the backups are being taken from every local etcd member on the same node. This avoids overwhelming a specific member with backup requests.

Since a Backup pod exist on every master node, and a `hostPath` volume is being used to store the backups, it is guaranteed to have available backups on all master nodes, which is required for disaster recovery scenarios.

_**Cons**_

No issue has been encountered during testing this approach.

> ### Open Questions

Author: has been decided above to 

- default schedule once at mid night `0 0 * * *`
- default retention to be the last 3 backups

* Backups are taken according to a default schedule.
* Backups are being maintained according to a default retention policy.

### Non-Goals

* Save cluster backups to remote cloud storage (e.g. S3 Bucket).
  - This could be a future enhancement or extension to the API.
  - Having backups in an independent storage from OCP cluster is crucial for Disaster Recovery scenarios.
* Automate cluster restoration.
* Provide automated backups for non-self hosted architectures like Hypershift.

## Proposal

This work utilizes the [API](https://github.com/openshift/enhancements/blob/master/enhancements/etcd/automated-backups.md#api-extensions) from current `CronJob` based implementation.
Moreover, it uses the same [PeriodicBackupController](https://github.com/openshift/cluster-etcd-operator/blob/master/pkg/operator/periodicbackupcontroller/periodicbackupcontroller.go#L81) as this proposal builds upon it, and also under the same feature gate.
The default configuration is created through `config.openshift.io/v1alpha1 Backup` CR, with name `default`, to distinguish it from other user supplied `Backup` CRs.
Using a CR enable users to override the default backup configuration.

> By design, the _**No-Config**_ backup approach works _orthogonal_ to the [Cron](https://github.com/openshift/enhancements/blob/master/enhancements/etcd/automated-backups) based method, and not a replacement.

Reviewer 7: Before anything else we need to expand on this section to mention how exactly are the `automated local backup` and `automated PVC backup` orthogonal or related to each other. 
How does one get enabled and why they need to share the same CRD?

Elsewhere you've mentioned:
>However, creating a `Backup` CR with `name=default` enables the container.

So that means the local backups get enabled when you have a special name i.e `default`? The name isn't very indicative of how this chooses between the `local` vs `PV` storage location.

And the `Backup` CR also has a field for `PVCName`. So is it that just getting ignored if the user makes edits to that when the name is `default`? Or is it immutable in that case? 
And then if you create another `Backup` CR with `name=foobar` does that enable the automated backups to PVC feature?

So now you have 2 CRs of the same type/spec with different names for different behaviors? Seems a bit unintuitive and obscure to me. 

Which leads me to the question on whether or not both these features need to share the same CR.
If they do, then what are the possible locations where the backups are being saved 
1. Locally across all nodes
2. On the PVC 
3. On both at the same time, or exclusively on 1 or 2.

Instead of relying on a magic name to enable the `local` option can we not explicitly have an API field(or multiple) to make it clear what location is being picked and how that can be changed.

And if there are drawbacks to having a unified API for both options  then we should list them out, and explore what having an entirely separate CR for the `automated local backup` would look like to truly make it orthogonal.

Author: > Before anything else we need to expand on this section to mention how exactly are the automated local backup and automated PVC backup orthogonal or related to each other.
How does one get enabled and why do they need to share the same CRD if they're orthogonal?

Added 👍🏽

Author: > So that means the local backups get enabled when you have a special name i.e default? The name isn't very indicative of how this chooses between the local vs PV storage location.

So the `default` name has been selected by @tjungblu since the very beginning, and we agreed to use it as indicator for the `NoConfig` method 👍🏽

Author: > And the Backup CR also has a field for PVCName. So is that just getting ignored if the user makes edits to that when the name is default? Or is it immutable in that case?

Yes it is being ignored, it is never mounted see https://docs.google.com/document/d/1813mZo4KEfTWw_FRbmG8Un91oPA9-eF29FcrZk-RLxE/edit?usp=sharing

Author: > And then if you create another Backup CR with name=foobar does that enable the automated backups to PVC feature?
So now you have 2 CRs of the same type/spec with different names for different behaviors? Seems a bit unintuitive and obscure to me.

Creating any `Backup` CR with any other name than `default`, triggers the `Cron` based approach

Only CR with name `default` is being handled by the NoConfig 

see test here https://docs.google.com/document/d/1Ee9a7GJsae8EA7komPcEV72YlRgbSbR0B0KSo3XRBUc/edit?tab=t.0#heading=h.hyvb3n32a77z 

Also applying two consecutive Backup CRs with name = `default`, the latest CR is being used to configure the DaemonSet and the Pods

Author: > Instead of relying on a magic name to enable the local vs pvc option can we not explicitly have an API field(or multiple) to make it clear what location is being picked and how that can be changed.

Well, I think the purpose of this work, is to avoid having user's configuration. In this sense, adding an `API` while the user should not use it ?!!! 🤔

Reviewer 6: I think we may want to get the three of us together to have a conversation about this project. 

One major question that still exists for me is, why is this not just a special case of the periodic backup.

It strikes me, that the CEO could enforce creating a `configv1.Backup` (from the periodic EP) that configures backing up to a hostpath on the master instances, and, prevent users from deleting or tweaking that in ways we don't want them to with admission time checks.

The default backup object that CEO creates would configure these default backups, but leverage the same cronjob based implementation as the rest of the backups.

Meaning that all that is needed for this project, is to create and work out the caveats of creating that default backup object, and not have to completely reimplement a way to handle the backups?

What about the existing periodic backup means it's not compatible in this way?

Author: > It strikes me, that the CEO could enforce creating a configv1.Backup (from the periodic EP) that configures backing up to a hostpath on the master instances, and, prevent users from deleting or tweaking that in ways we don't want them to with admission time checks.

Actually user could change the default configs by editing the `default` Backup CR.

> The default backup object that CEO creates would configure these default backups, but leverage the same cronjob based implementation as the rest of the backups.

I actually think the main reason we have changed the approach is to have replicas across all master nodes iiuc

> What about the existing periodic backup means it's not compatible in this way?

I think the purpose since the beginning was to have both approaches working simultaneously iiuc

> I think we may want to get the three of us together to have a conversation about this project.

I am fine, I will work out an invite for this 👍🏽

Reviewer 6: > I actually think the main reason we have changed the approach is to have replicas across all master nodes iiuc

Currently, what does a periodic backup do? Just run the one replica on any node? Or does it run one replica on any master node?

> I think the purpose since the beginning was to have both approaches working simultaneously iiuc

I don't see why having multiple periodic backups configured would mean that there's a non simultaneous element

Author: > Currently, what does a periodic backup do? Just run the one replica on any node? Or does it run one replica on any master node?

I think it goes to any `master node` 

cc @tjungblu @hasbro17

Author: > I don't see why having multiple periodic backups configured would mean that there's a non simultaneous element

What I meant, is that both approaches exist :)

During [testing](https://github.com/openshift/cluster-etcd-operator/pull/1354#issuecomment-2408661261), this method has shown a solid and stable deployment, where the Backup pods shown only `1` restart over a duration of `45` minutes. Moreover, the cause of the restart on all the Backup pods were due to `node` issue, not the backup container.

_**Pros**_

As the Backup pods are being scheduled independently of the etcd pods, any error within the Backup pods does not influence cluster Openshift cluster availability.

Moreover, as the Backup pods are being scheduled within every master nodes, the backups are being taken from every local etcd member on the same node. This avoids overwhelming a specific member with backup requests.

Since a Backup pod exist on every master node, and a `hostPath` volume is being used to store the backups, it is guaranteed to have available backups on all master nodes, which is required for disaster recovery scenarios.

_**Cons**_

No issue has been encountered during testing this approach.

### Open Questions

> - What values could be used for default configurations, considering all OCP variants.

Reviewer 4: Not sure which "values" this is about - requests/limits?

Reviewer 5: How about a default interval such that every node has a backup from the last 24h and retention being three backups per node in case there's a need to find a backup from before a particular change in the last 72h.

Author: I actually used this `00:00` as default value 

See https://github.com/openshift/cluster-etcd-operator/pull/1354/files#r1882979478 

here is the [commit](https://github.com/openshift/cluster-etcd-operator/pull/1354/commits/564eb9ac3667ae9801c9276b3ac251f1d243be2b) fixing this [bug](https://issues.redhat.com/browse/OCPBUGS-43687) as reported by @sandeepknd

Author: need consensus from everyone 

cc @openshift/openshift-team-etcd  @vrutkovs @JoelSpeed

Reviewer 6: David's suggestion seems reasonable.

Did we conclude on the thread about taking an initial backup when the node is first created? (to avoid CPMS rollout clearing all backups and then needing a backup between the rollout and 00:00)

Author: I think it is best to discuss during the control plane meeting, so we get input from everyone 

Wdyt @JoelSpeed @deads2k @tjungblu  ?

Author: > Did we conclude on the thread about taking an initial backup when the node is first created? (to avoid CPMS rollout clearing all backups and then needing a backup between the rollout and 00:00)

I think we can add this to GA requirement, wdyt @deads2k

Reviewer 6: > iirc, the new nodes created by CPMSO does not become running till the etcd pod on the node not only joins the etcd cluster, but also becomes voting member

This is not true, CEO and etcd have nothing to do with the Node becoming ready, or the Machine moving into the Running phase.

The CEO adds a machine deletion hook to the machine, when it intends to promote an etcd member to voting. This is part of the machine spec, and so you couldn't gate scheduling on it using normal scheduler mechanics.

Author: Ok, what do you think if CEO labels machines according to its role in etcd cluster (i.e. learner, voting) ?

Reviewer 6: If CEO were to add a label to the nodes, which stated whether they were voting or learner members, then yes, that could be used as a scheduling gate to make sure backups aren't taken until members are voting.

Another part of this thread, which I think we need to discuss wider, is whether we are concerned about CPMS rollouts and this mechanism.

Eg, if we are backup up daily at midnight, and I rollout CPMS at around 8am, there's a potentially long period where there will be no backups (CPMS deleting last old machine to midnight). And are we concerned about this scenario?

We could argue that for CPMS users they should use an external backup, and this mechanism is best effort, but lets make sure we have stakeholder agreement on that.

The alternative is to try and come up with a way to schedule backups each time a member is promoted, which would mean we _should_ always have at least 1 backup

Reviewer 6: > How about a default interval such that every node has a backup from the last 24h and retention being three backups per node in case there's a need to find a backup from before a particular change in the last 72h.

I'm ok with this, but we should get agreement about what happens when CPMS wipes this state. Perhaps discussing on the control plane arch on tuesday is worthwhile?

Author: @JoelSpeed  I raised this [commit](https://github.com/openshift/cluster-etcd-operator/pull/1354/commits/7706a916815d563bdb1ba4c036e53d3c462a513b) .. Here is the changes 

- List All Etcd Voting Members
- List All Master Nodes
- Use All Master Node with Etcd Voting Member
- Apply a `votingNode` label to all of them to be used by the DS for scheduling
- Upon removing the `default` Backup CR, delete `votingNode` label from all master nodes

creating a cluster to see it in action now

Reviewer: CEO should not touch the node object, if anything we would keep it in the backup CRD status somehow

Author: @tjungblu Why not 🤔 

> should not touch the node object

`clustermembercontroller` and `clustermemberremovelcontroller` within CEO works mainly on the node objects 🤔 

Also if you check the [implementation](https://github.com/openshift/cluster-etcd-operator/pull/1354/commits/7706a916815d563bdb1ba4c036e53d3c462a513b), you will find that I add the label upon applying the CR and removing the labels upon deleting the CR 

please correct me if I'm wrong 👍🏽

> - What is the minimum disk size needed in order to operate this feature without running out of disk space ?

Reviewer 4: No less than current DB size (assuming that compression won't increase it)

Author: yes I think we need an alert for this, this is being tracked here https://issues.redhat.com/browse/ETCD-692

> - How to handle an empty `schedule` ? 

Reviewer 4: I'd say empty schedule is invalid - its better be explicit than implicitly set "@weekly" or smth. Can set this to "weekly" default, but disallow unsetting it?

Author: alright, so QA found this bug https://issues.redhat.com/browse/OCPBUGS-43687 

There were two test scenarios 

- schedule: "" 
```
Fails with error:
The Backup "default" is invalid: spec.etcd.schedule: Invalid value: "": spec.etcd.schedule in body should match '^(@(annually|yearly|monthly|weekly|daily|hourly))|(\*|(?:\*|(?:[0-9]|(?:[1-5][0-9])))\/(?:[0-9]|(?:[1-5][0-9]))|(?:[0-9]|(?:[1-5][0-9]))(?:(?:\-[0-9]|\-
```
     
- schedule is being omitted completely from the CR 
```
$ oc logs backup-server-daemon-set-m96bv
F1022 13:26:10.239742       1 backupserver.go:77] error parsing backup schedule : empty spec string
```

I raised a fix for the second case, since the first is handled by the API-server 

https://github.com/openshift/cluster-etcd-operator/pull/1354/commits/564eb9ac3667ae9801c9276b3ac251f1d243be2b

Author: The fix uses default schedule which is 

```
defaultBackupSchedule = "0 0 * * *"
```

daily midnight

> - How to mitigate losing backups upon control-plane scaling using `CPMSO` ?

Reviewer 4: I don't see why should we should worry about this - if the backups are interchangable any other backup would do. If they are not you won't be able to restore this node anyway

Reviewer 6: If something goes drastically wrong during a CPMS roll out and you end up in a position where you've lost all your backups, you'll probably be pretty annoyed.

If the backup pods are running as daemonsets and running always, perhaps they should check on startup for the presence of any backups, and, produce a backup shortly after starting if nothing else is present?

That would hopefully mean that by the time the CPMS has finished rolling the second node, the first new node will have at least one backup, so there is at least some backup somewhere

Author: > If the backup pods are running as daemonsets and running always, perhaps they should check on startup for the presence of any backups, and, produce a backup shortly after starting if nothing else is present?

I think producing a backup from a fresh node, should take place only in either of two cases are valid 

- the etcd member on the new node is no longer `learner` (i.e. it is a voting member of the etcd cluster)
- the backup pod gets a backup copy from other node

I think the latter is  unnecessary complicated, and the first is the most straightforward solution 

That being said, I think this case could be handled best using backups stored on `External Targets` (e.g. S3 bucket)

Author: raised this to track https://issues.redhat.com/browse/ETCD-694

feel free to comment to shape the requirements

Reviewer 6: Seems reasonable to wait for the member to be voting, +1 to that

And you're correct, the external backup is the ultimate solution, but, if we did a one off backup straight after the etcd member became voting, that would make this feature, in combination with CPMS, much more reliable

Author: @deads2k I need your opinion here, what to do with a new member joining the cluster ? 

see above

Author: good point 👍🏽 

So the fix I created [here](https://github.com/openshift/cluster-etcd-operator/pull/1354/commits/3f289fc0fb7560af8582b780c454d99d9960a04b) uses `etcdcli` to list all the `voting` members, then maps the members to master nodes. This guarantees to schedule only to master nodes with `voting` etcd member

I think if I use the endpoint approach, it would be way easier 👍🏽

Author: raised this [fix](https://github.com/openshift/cluster-etcd-operator/pull/1354/commits/5403ca0c1e1163ea6d8c94d35e433ea34e2627c1) to use the endpoint-cm instead of  etcd voting members mapping 👍🏽 

thanks @tjungblu 🙏🏽

Author: @tjungblu tested the implementation using Etcd-endpoints CM on OCP 4.19 cluster 

result is here https://docs.google.com/document/d/14npia5cM3E1G-k-8KiWyh34PcfGDVx6YZIvJhLKo1Y0/edit?usp=sharing

> 

Reviewer: alternative approach we just discussed on the call:
* have a no-config backup controller inside of CEO running the delay queue
* at every interval create backup CRs for each control plane node

that way we have maximum reuse of all components. We have to add a node selector to our current backup CR however.

Reviewer 6: To allow configuration of this option, could the `PeriodicBackup` still be used, but have a `hostPath` option that triggers this mode of operation.

Once concern with having no PeriodicBackup at the higher level would be that the user has no way to adjust the default backup configuration should they want to

Author: raised https://github.com/openshift/cluster-etcd-operator/pull/1363 yesterday to test the approach proposed above

- Used parallelism in JobSpec = 3 (i.e. control-plane nodes)
- Used `NodeAffinity` in order to schedule each `Job` on  master nodes only
- Used `PodAntiAffinity` in order to guarantee that each `Job` is being scheduled on a different master node

However, the approach did **not** work for the following reasons 

- But the Backup [CR](https://github.com/openshift/api/blob/684b2b1679baeb7742ec7b76c6eab8e0c8036e56/config/v1alpha1/types_backup.go#L99) expects a PVC per CR, this is not possible here, since we need 3 different PVCs, one per each node, before hand, and the CR expects only one PVC
- following a suggestion [here](https://github.com/openshift/cluster-etcd-operator/pull/1363#issuecomment-2444641024), I tried to create a `hostPath` volume within each `Jop.Spec` from within `PeriodicBackupController` to overcome the PVC singleton issue, however, the `request-backup` and `prune` containers that each `Backup` job consist of, expect `PVC` as an arg.

Reviewer 6: And how is this different then to the daemonset approach? Is the daemonset approach using a different backing up binary?

Did the pods schedule as expected, even if they weren't able to proceed due to PVC mounting issues?

Author: > And how is this different then to the daemonset approach? Is the daemonset approach using a different backing up binary?

So the DS approach uses a separate binary (i.e. `etcd-backup-server`) see https://github.com/openshift/cluster-etcd-operator/pull/1354 

The `etcd-backup-server` expects all the `Backup` CR as args .. also it uses default path with `hostPath` volume to store backups 

> Did the pods schedule as expected, even if they weren't able to proceed due to PVC mounting issues?

So the pods were not scheduled since the PVC were not mounted, the pods could not become `Ready`

### Implementation Details/Notes/Constraints

The _no-config_ backups builds atop of the existing [automated backup of etcd](https://docs.openshift.com/container-platform/4.15/backup_and_restore/control_plane_backup_and_restore/backing-up-etcd.html#creating-automated-etcd-backups_backup-etcd).

The `PeriodicBackupController` within **cluster-etcd-operator** checks for `Backup` CR with `name=default` within every sync. Upon applying a `Backup` CR with `name=default`, the _no-config_ backup sidecar container is being enabled within each etcd pod in Openshift cluster.

> The `Backup` configurations such as `scheudle`, `timezone` and `retention policy`, are being supplied to the _no-config_ container as flags. 

Reviewer 4: ```suggestion
The `Backup` configurations such as `schedule`, `timezone` and `retention policy`, are being supplied to the _no-config_ container as flags. 
```

reviewers:
  - "@JoelSpeed"
  - "@dusk125"
  - "@hasbro17"
  - "@tjungblu"
  - "@vrutkovs"
approvers:
  - "@hasbro17"
api-approvers:
  - "@JoelSpeed"
creation-date: 2024-06-17
last-updated: 2024-10-21
tracking-link:
  - https://issues.redhat.com/browse/ETCD-609
see-also:
  - "https://issues.redhat.com/browse/OCPSTRAT-1411"
  - "https://issues.redhat.com/browse/OCPSTRAT-529"
---

# Automated Backups of etcd with No Config

## Summary

To enable automated backups of Etcd database and cluster resources of an Openshift cluster from day one after installation, 
without provided configuration from the user.

This document outlines the possible approaches that were investigated, discussed and tested, in addition to the final implemented approach.  

## Motivation

The current method of **Backup** and **Restore** of an Openshift cluster is outlined in  [Automated Backups of etcd](https://github.com/openshift/enhancements/blob/master/enhancements/etcd/automated-backups.md). 
This approach relies on user supplied configuration, in the form of `config.openshift.io/v1alpha1 Backup` CR. This method utilizes `batchv1.CronJob` in order to take Backups of Etcd and cluster resources.

> To improve the user's experience, this work proposes taking etcd backup using default configurations from day one, without user interference.

Reviewer 9: s/interference/intervention/

Moreover, for recovery in disaster scenarios, where more than one control-plane node is lost from an Openshift cluster, this proposal attempts to store backups on all control-plane nodes.
Having backups of an Openshift cluster on all control-plane nodes, improves the recovery probability in disaster scenarios.

### User Stories

* As a cluster administrator, I would like an OCP cluster backups to be taken without user's supplied configuration from day one, after installation.
* As a cluster administrator, I would like to an OCP cluster backups available on all master nodes, so that I have a recent cluster state to recover from in the event of losing control-plane nodes.

### Goals

* Backups should be taken without configuration after cluster installation from day 1.
* Backups are saved to a default location on each master node's file system.
* Backups are taken according to a default schedule.
* Backups are being maintained according to a default retention policy.

### Non-Goals

* Save cluster backups to remote cloud storage (e.g. S3 Bucket).
  - This could be a future enhancement or extension to the API.

>   - Having backups in an independent storage from OCP cluster is crucial for Disaster Recovery scenarios.

Reviewer 8: [Our current restore docs][1] list:

> A healthy control plane host to use as the recovery host.

as a prerequisite.  Possibly having backups on out-of-cluster storage is useful for audit purposes.  And it may be useful for some future recover-from-nothing procedure.  But at the moment, I don't see how it would be critical for Disaster Recovery.  Which is good, because if a non-goal was critical, that would make it pretty unclear why it was worth doing the enhancement at all ;)  Can we drop this line?

[1]: https://docs.openshift.com/container-platform/4.17/backup_and_restore/control_plane_backup_and_restore/disaster_recovery/scenario-2-restoring-cluster-state.html#dr-scenario-2-restoring-cluster-state_dr-restoring-cluster-state

Author: `lgtm` 🙏🏽 🙏🏽 🙏🏽 🙏🏽 🙏🏽

Reviewer: we already have this functionality discussed in the other EP, hence it's out of scope in here

Author: done 👍🏽

* Automate cluster restoration.

> * Provide automated backups for non-self hosted architectures like Hypershift.

Reviewer 8: I'm late to the party here, but it's probably worth some text on this exclusion down in the:

> ```markdown
> #### Hypershift / Hosted Control Planes
> This proposal does not affect HyperShift.
> ```

section.  Because while it's possible that you can implement for standalone, and then do something similar for HyperShift (or MicroShift), those folks will still _want_ automated backups eventually, right?  And it's good to at least have a rough idea of how automated backups would work on those systems.  Would you use a completely different type?  Would you stick with `Backup.config.openshift.io`, and drop your current `+genclient:nonNamespaced` and have `pvcName` be required on HyperShift?  Some kind of plan.

Author: great point, wdyt  @sjenning 🤔

Reviewer 9: Yes, hypershift does not currently run the etcd-operator so this would not apply to it

Reviewer 10: We do want to eventually have automatic backups in HyperShift, but agree that we would want a different API possibly within the HostedCluster CR.

One thing that I would like to see called out here though is that because the Backup CRD is not actually useful in a HyperShift hosted cluster, it should not be applied by the CVO to clusters with an external control plane.

## Proposal

This work utilizes the [API](https://github.com/openshift/enhancements/blob/master/enhancements/etcd/automated-backups.md#api-extensions) from current `CronJob` based implementation.

> Moreover, it uses the same [PeriodicBackupController](https://github.com/openshift/cluster-etcd-operator/blob/master/pkg/operator/periodicbackupcontroller/periodicbackupcontroller.go#L81) as this proposal builds upon it, and also under the same feature gate.

Reviewer 6: You may want to consider gating this separately so that you don't tie the cronjob based backups to this implementation, delaying their GA

Author: +1 @tjungblu, 

we have discussed this above, and as it is the same GA feature, we should not use a diff FeatureGate 

see https://github.com/openshift/enhancements/pull/1646#discussion_r1654864573

The default configuration is created through `config.openshift.io/v1alpha1 Backup` CR, with name `default`, to distinguish it from other user supplied `Backup` CRs.

> Using a CR enable users to override the default backup configuration.

Reviewer 10: Does *not* having a CR named `default` result in a default configuration? If we plan to have a default CR, can that CR be deleted? What if I enable the CronJob backups featurre but not the default backups, will I be prevented from creating a CR named 'default'?

Reviewer 10: I don't know that it will be immediately obvious that a CR with a special name should behave differently than other CRs of the same type.
IMO it might make sense to create a separate configuration CRD for this (could share most of its spec with the current CRD)

Reviewer: good points, Cesar. I think I was driven mostly by the special case "default" namespace, which is a similar kind of special case.

Reviewer 6: If we use the installer to generate the default Backup CR, and then have it applied during cluster bootstrap. That removes any logic about it being a special name, and allows users to delete and remove it should they wish.

I believe the current approach however is a create only CR inside the payload, which means if you deleted the CR, it would come back

Author: >  What if I enable the CronJob backups featurre but not the default backups, will I be prevented from creating a CR named 'default'?

So this has been tested within this Spike https://docs.google.com/document/d/1Ee9a7GJsae8EA7komPcEV72YlRgbSbR0B0KSo3XRBUc/edit?usp=sharing 

As you can see, both approaches could be running the same cluster simultaneously 

> IMO it might make sense to create a separate configuration CRD for this (could share most of its spec with the current CRD)

I believe if we were to create another CRD, it will be just a clone of the current [Backup](https://github.com/openshift/api/blob/188620a9ab5eaad36199cf83d9a0afcb08fdfd52/config/v1alpha1/types_backup.go#L19) type. 

I don't think this is correct, to have duplicate API types, or 🤔

> By design, the **No-Config** backup approach works alongside to the [Cron](https://github.com/openshift/enhancements/blob/master/enhancements/etcd/automated-backups) based method, and not a replacement, as shown in this [test](https://docs.google.com/document/d/1Ee9a7GJsae8EA7komPcEV72YlRgbSbR0B0KSo3XRBUc).

Reviewer 10: Should you link to a doc that not everybody can see?

Author: indeed, the doc is being accessible to anyone in RedHat https://docs.google.com/document/d/1Ee9a7GJsae8EA7komPcEV72YlRgbSbR0B0KSo3XRBUc/edit?usp=sharing 

maybe you used your personal account ? 🤔

Reviewer: I think Cesar says that this doc should be available for everyone in public or drop the link entirely

Author: updated to public doc 👍🏽

Although both approaches use the same `config.openshift.io/v1alpha1 Backup` CR for configuration, the No-Config method does not utilise the `PVC` field in `Backup` CR. 

> In fact, configuring the PVC field for the No-Config method is being ignored as this [test](https://docs.google.com/document/d/1813mZo4KEfTWw_FRbmG8Un91oPA9-eF29FcrZk-RLxE) shows. 

Reviewer 6: What does this mean? If a user were to tweak the default configuration, we wouldn't observe those changes? That seems odd

Author: so the NoConfig backup should **not** consider user provided configuration, including the PVC. 

The DaemonSet logic does not consider any configured PVC within the CR

Reviewer 6: That is a really weird distinction to make, and will likely cause confusion for end users. Why has this choice been made? What happens in your Job based POC? Does it observe the configuration?

Author: So this is just PoC, not the final code to be rolled out. 

If we were to choose this as in the Poll I created, then I would have better code. 

However, I really think the DS approach is better, specially that parallelizing the CronJob results into much more created resources (i.e. Job, Pod, Job, Pod) .. 

cc @deads2k @tjungblu

Reviewer 6: > However, I really think the DS approach is better, specially that parallelizing the CronJob results into much more created resources (i.e. Job, Pod, Job, Pod) ..

Lets continue a discussion about the pros and cons of both approaches somewhere else, perhaps synchronously

And ok, so this is the POC, but what do you want the end results to be? The EP should reflect the desired end state, and so, if you think the end result is that the changes would be ignored, you can write that here and we can discuss. If you think the desired end state is that end users changes to the config are observed (I would strongly encourage this), then that's what we should write down here

> Instead, **No-Config** backups uses a default `hostPath` location on all control-plane nodes. 

Reviewer 11: Might be a separate issue but should the hostPath location be specified in the Backup spec? Might save cluster administrators time looking up documentation on where to find the default backups.

Reviewer: we definitely will document it, I think the backup CR has the path in its status IIRC (we use it in e2e testing)

Author: +1 @tjungblu, yes it updates the CR status with the path of the backup

This ensures that backups are always accessible if any of the master node becomes unavailable.

> Using `hostPath` allows the **No-Config** to work on all variants of Openshift (e.g. Single-node and Bare-metal), as well as having backups on all control-plane nodes, for recovering from disaster scenarios.

Reviewer 11: Are there any security concerns creating a proliferation of `hostPath` volume mounts? I can recall security concerns with hostPath historically but I'm not up to date on if this is still a valid concern.

Secondly, can the default `hostPath` backups be disabled if desired, perhaps just by deleting the CR?

Reviewer: There isn't any other logical choice here. We need to store the backups locally on each node for easier access on restoration. We can't rely on LSO or other external PVCs when the control plane is down. 

It's still a valid concern though, we have to accept this as a trade-off due to the functional requirements.

Author: +1 @tjungblu 

> Are there any security concerns creating a proliferation of hostPath volume mounts? I can recall security concerns with hostPath historically but I'm not up to date on if this is still a valid concern.

yes indeed the hostPath gives access to the master node's file system. But as @tjungblu explained, we need to have access to the backups on master nodes. 

- LSO is not applicable on all OCP variants.
- `LocalVolume` are not accessible if the master node is unreachable/unavailable.
- Cloud-based Storage through PV is not available to all OCP variants as well. 

Hence, `hostPath` is the only option available. 

Ideally however, external target for storing backups is the ultimate solution.

To sum up, the main two differences between the Cron-based and No-Config backups method are on storage and replication.
The Cron-based stores the backups on the mounted PVC, while No-Config approach uses `hostPath` on all master nodes, regardless of any PVC configured on the `default` `config.openshift.io/v1alpha1 Backup` CR.

### API Extensions

This proposal uses the same [API](https://github.com/openshift/enhancements/blob/master/enhancements/etcd/automated-backups.md#api-extensions) from the Cron based method.

> No additional API changes are required.

Reviewer 6: I remember having conversation before about adding hostpath as an option to the existing API, so that this EP could leverage that config

Do you have an example of the CR as it would look for the default config?

Author: Yes, as @deads2k and yourself voted above, we can use the following configs 

- [here](https://github.com/openshift/enhancements/pull/1646#discussion_r1882929430)
- [here](https://github.com/openshift/enhancements/pull/1646#discussion_r1883744285)

```yaml
apiVersion: config.openshift.io/v1alpha1
kind: Backup
metadata:
  name: default
spec:
  etcd:
    schedule: "0 0 * * *"
    timeZone: "UTC"
    retentionPolicy:
      retentionType: RetentionNumber
      retentionNumber:
        maxNumberOfBackups: 3
```

Actually I have used the same values for fixing a bug raised by QA during testing  [here](https://github.com/openshift/cluster-etcd-operator/pull/1354/commits/e320c4befbc11fe93071a0a668538a67ea7dea1f)

Author: @JoelSpeed shall we resolve this ?

Reviewer 6: I'm not sure this is resolved no. If we are using a hostpath option, and typically this configuration would include a PVCName, I think we need to have an explicit choice between whether the user meant for hostpath, or, whether they intended to specify a PVC. How would you know if they forgot to specify the PVC vs wanted host path? I don't think that's possible if we just omit the PVC part. We want this to be a declarative and required decision on the end user part

Author: ok, how do you want to distinguish this then ? 

I used the same approach as the default schedule config (i.e. "0 0 * * *"). 

If the user were to override the schedule, then the user's schedule applies. 

Similarly, the default is HostPath. If the use were to override the by configuring a PVC, then the user's config applies

Please correct me if I'm  wrong 🤔

Reviewer 6: Create a DU on this API that is explicit and required, make the user make a choice, do not be implicit 

```
storage:
  type: PVC | HostPath
  pvc:
    name: ...
```

> 

Reviewer 10: What about status? How do I know that my no-config etcd backups are happening?

Author: So till now I am using the `periodicBackupController`. If the CR `name==default`, then this initiate DaemonSet 

The status is being updated accordingly 

```
		if len(terminationReasons) > 0 {
			_, _, updateErr := v1helpers.UpdateStatus(ctx, c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
				Type:    "PeriodicBackupControllerDegraded",
				Status:  operatorv1.ConditionTrue,
				Reason:  "Error",
				Message: fmt.Sprintf("found default backup errors: %s", strings.Join(terminationReasons, " ,")),
			}))
			if updateErr != nil {
				klog.V(4).Infof("PeriodicBackupController error during [etcd-backup-server] UpdateStatus: %v", err)
			}
			return nil
		}
	}
```

Author: @csrwng would this be enough to resolve this ?! 🤔

TBD
### Workflow Description
TBD

## Design Details

To back up the Etcd datastore and cluster resources of an Openshift cluster for the purpose of restoration, a persistent storage must be utilised.
Several main point have been considered during designing the implementation of this proposal, however the storage solution to be used is the most influential, particularly to be used by all Openshift variants.  

### Storage

Ideally, the storage solution to use should have the following characteristics :-

- Backups should be taken on each master node to avoid overwhelming a specific etcd member more than others.
- Spreading the backups among all etcd cluster members provide guarantees for disaster recovery in case of losing two members at the same time.
- One fits all storage solution, to be used across all Openshift variants including Bare metal and Single Node Openshift.

According to these requirement, only `hostPath` and `localVolume` storage could be utilized. 
However, for `localVolume` storage, if the master node where the `PV` is mounted became unhealthy/unavailable/unreachable, then the backups are no longer accessible, also taking new backups is no longer possible.

> Hence, `hostPath` has been used as it is the only viable option that could be utilized on all Openshift variants. 

Reviewer 5: good explanation, thank you.

_**Cons**_

Etcd pod is a static-pod, and the availability of the etcd cluster is vital for cluster stability. As the Etcd cluster is being managed by the `Cluster-etcd-operator`, and a `Deployment` resource manages the _mirrored_ etcd pods, any changes in the pod manifest results into rolling out a new revision, which leads to a duration of `etcd-cluster` unavailability, till the new revision is rolled out for all etcd pods. Consequently, the `API-Server` can not process any write requests, and read requests are being handled; if any, from the API-Server cache.   

The backup sidecar container is being `disabled` by default. Upon creating a `Backup` CR on an Openshift cluster, the `PeriodicBackupController` reacts by enabling the container. This leads to a change in etcd static pod manifest, and a new revision is being rolled out. As explained above, such action causes etcd-cluster and API-Server unavailability.

Additionally, any error within the backup sidecar, the container goes into `CrashLoopBackOff`, which leads to the whole `Pod` reports `NotReady` status. Consequently, the whole `Etcd-cluster` can not server requests, nor the `API-Server`.

> #### DaemonSet manages an Etcd-backup pod within every control-plane node.

Reviewer 6: We had some previous discussion about using the parallel execution semantics of cronjobs, how did you get on with that research activity?

Author: yes, I have tested it here https://github.com/openshift/cluster-etcd-operator/pull/1363

Author: Yes, I have already tested this here https://github.com/openshift/cluster-etcd-operator/pull/1363 

you also have commented on the PR :D :D

Reviewer 6: Ok, do we need to update the details here with the results of that investigation then?

Reviewer 6: Right, just checking that, I don't really see why it was dismissed?

My understanding of the situation we had was:
* You tried to use the cronjob with parallelism
* You got stuck because you needed a PVC
* I suggested you just change it to use hostpath for now
* ...

Somewhere on this EP we discussed adding a hostpath option as an alternative configuration for PVC, and I believe I was told that this was on the roadmap for the backup feature anyway?

If you were to use the cronjob approach with hostpath, and the parallelism and completions config, would that work to schedule the backups, and have them on the hostpath, without the need for creating this whole new DS approach?

Author: Well, the ultimate  goal of this work is to provide AutomatedBackup **without** user supplied configuration.

I recall during testing https://github.com/openshift/cluster-etcd-operator/pull/1363 on a live cluster, I got stuck with configuring PVC, running parallel jobs requires parallel PVC 

I also tried patching the scheduled pods with hostPath but I don't remember why it did not work out 

wdyt @deads2k @tjungblu @vrutkovs  ?

Reviewer: if we use a hostPath (and we finally all agree) - then it might make sense to have a dedicated field for it

Reviewer 6: I have suggested with @Elbehery privately, that we move from having 
```
spec:
  ... // Other fields
  pvcName: ...
```
To
```
spec:
  ... // Other fields
  storage:
    type: PVC | HostPath
    pvc: // Only valid when type is PVC, required when type is PVC
      name: ...
```
For the automated backup case, we can create the default Backup with the storage type as HostPath, and the operator can modify the way it makes the podSpec for the cronjob based on the storage type. That would allow us to avoid the "we need multiple PVCs" option, but also allow, should the end user desire, to reconfigure the default, automated backup away from using hostpath and to an off cluster PVC storage, at which point we go back to 1 execution of the job at a time, as with any other backup

Reviewer 12: Why is `type` then needed anyway? Can't you just discriminate based on whether the key `pvc` or `hostPath` exists? This is less error-prone (no mismatch possible); it only needs to be checked that not both are provided at the same time. (sorry, just stumbled over this, please ignore my comment if it doesn't make sense semantically; my comment is only based on API syntax)

Reviewer 6: We use [Discriminated Unions](https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#discriminated-unions) so that we have a clear intention from the end user, which of the options they intended. 

Non-discriminated unions in upstream kube have caused confusion over time, and now the convention is to lean towards discriminated unions. More verbose, but more intentional.

> This is less error-prone (no mismatch possible)

We use CEL to ensure that the type, and non-empty union member are paired, so, there's no way for an end user to persist `pvc: {}` with `type: HostPath`.

Reviewer 12: Thanks. I didn't know that there was already a convention about this. Personally, I think stating the top-level key is intention "enough" for non-squashed configuration groups, but I guess there has already been tons of discussion around this, so I for sure don't question the consensus here.

Sorry for my ignorance and the noise ...

Author: > I have suggested with @Elbehery privately, that we move from having

@JoelSpeed  The result of this test is [here](https://github.com/openshift/enhancements/pull/1646#issuecomment-2564491734) 

So mainly, I programmatically patched the PVC to be with HostPath. This is the same outcome if we were to change the API as you described above 👍🏽

Reviewer 6: Test looks positive for me, if we decide that is the path to take, we can update the API as we progress

- How to handle an empty `schedule` ? 
- How to mitigate losing backups upon control-plane scaling using `CPMSO` ?

### Implementation Details/Notes/Constraints

The _no-config_ backups builds atop of the existing [automated backup of etcd](https://docs.openshift.com/container-platform/4.15/backup_and_restore/control_plane_backup_and_restore/backing-up-etcd.html#creating-automated-etcd-backups_backup-etcd).

The `PeriodicBackupController` within **cluster-etcd-operator** checks for `Backup` CR with `name=default` within every sync. Upon applying a `Backup` CR with `name=default`, the _no-config_ backup sidecar container is being enabled within each etcd pod in Openshift cluster.
The `Backup` configurations such as `schedule`, `timezone` and `retention policy`, are being supplied to the _no-config_ container as flags. 
According to user's preference, the default configuration could be modified by updating the CR.

The backups are being stored on the local disk of each master node independently, using `hostPath` volume. This replication is required for disaster recovery scenarios where more than one of the master nodes has been lost.

A `Backup` CR with default configuration below.

> ```yaml

Reviewer 6: Does this require any API changes to the existing API, is there a PR available for that?

Author: > Does this require any API changes to the existing API, is there a PR available for that?

No, no changes required, the `PVC` is Optional, and the QA. I have done all my tests using this CR 

You can find links to the test outputs below 

- https://github.com/openshift/cluster-etcd-operator/pull/1354#issuecomment-2408661261
- https://github.com/openshift/cluster-etcd-operator/pull/1354#issuecomment-2408662693
- [ETCD-691](https://docs.google.com/document/d/1813mZo4KEfTWw_FRbmG8Un91oPA9-eF29FcrZk-RLxE/edit?usp=sharing)
- [ETCD-690](https://docs.google.com/document/d/1Ee9a7GJsae8EA7komPcEV72YlRgbSbR0B0KSo3XRBUc/edit?usp=sharing)
- https://github.com/openshift/origin/pull/29187#issuecomment-2410826761
- https://github.com/openshift/cluster-etcd-operator/pull/1325/commits/85ca0a1a5219780d7b9077a354d33ece30e45743
- https://github.com/openshift/origin/pull/29187#issuecomment-2410826761

Reviewer 6: Looking at the other parts of this conversation, I think we want to have an explicit storage choice from the user. We need to avoid the user forgetting to add the PVC and ending up with hostpath for example. 

I think we should iterate on the API and create a storage DU here as well, and make that a required part. User either chooses PVC or HostPath, and it's an explicit choice

  - "@gnufied"
approvers:
  - "@deads2k"
  - "@JoelSpeed"
  - "@hasbro17"
  - "@tjungblu"
  - "@vrutkovs"
  - "@JoelSpeed"
api-approvers:
  - "@JoelSpeed"
creation-date: 2024-06-17
last-updated: 2024-10-21
tracking-link:
  - https://issues.redhat.com/browse/ETCD-609
see-also:
  - "https://issues.redhat.com/browse/OCPSTRAT-1411"
  - "https://issues.redhat.com/browse/OCPSTRAT-529"
---

# Automated Backups of etcd with No Config

## Summary

To enable automated backups of Etcd database and cluster resources of an Openshift cluster from day one after installation, 
without provided configuration from the user.

This document outlines the possible approaches that were investigated, discussed and tested, in addition to the final implemented approach.  

## Motivation

The current method of **Backup** and **Restore** of an Openshift cluster is outlined in  [Automated Backups of etcd](https://github.com/openshift/enhancements/blob/master/enhancements/etcd/automated-backups.md). 
This approach relies on user supplied configuration, in the form of `config.openshift.io/v1alpha1 Backup` CR. This method utilizes `batchv1.CronJob` in order to take Backups of Etcd and cluster resources.

> To improve the user's experience, this work proposes taking etcd backup using default configurations from day one, without user intervention.

Reviewer 13: question: from a support perspective does this mean it is a bug if the backup is invalid or corrupt? are there any legal concerns with Red Hat "owning" backups?

Moreover, for recovery in disaster scenarios, where more than one control-plane node is lost from an Openshift cluster, this proposal attempts to store backups on all control-plane nodes.
Having backups of an Openshift cluster on all control-plane nodes, improves the recovery probability in disaster scenarios.

### User Stories

* As a cluster administrator, I would like an OCP cluster backups to be taken without user's supplied configuration from day one, after installation.
* As a cluster administrator, I would like to an OCP cluster backups available on all master nodes, so that I have a recent cluster state to recover from in the event of losing control-plane nodes.

### Goals

* Backups should be taken without configuration after cluster installation from day 1.
* Backups are saved to a default location on each master node's file system.
* Backups are taken according to a default schedule.

> * Backups are being maintained according to a default retention policy.

Reviewer 13: Would it also be a goal to ensure that the backup is valid? Is there a periodic job that ensures this or is the cluster/customer trusting it will work?

Reviewer 13: I believe the size of the etcd db is no longer capped. Is this a concern for taking snapshots?

The advantage of using `DaemonSet` over Deployment is to guarantee scheduling etcd-backup-server pods on all `Control-plane` nodes, regardless of the number of nodes (e.g. 1, 3, 5) using `nodeSelector` capability.
By setting the `nodeSelector` to the master node label (i.e. `node-role.kubernetes.io/master`), etcd-backup-server's pods are guaranteed to scheduled on every control-plane node.

This method has been implemented and tested. The same container used in the Sidecar approach has been utilised. However, the scheduling is being managed by a DaemonSet.
Upon creating a `Backup` CR on an Openshift cluster, the `PeriodicBackupController` reacts by creating a DaemonSet resource, with the `etcd-backup-server` container and the configurations applied in the CR.
During [testing](https://github.com/openshift/cluster-etcd-operator/pull/1354#issuecomment-2408661261), this method has shown a solid and stable deployment, where the Backup pods shown only `1` restart over a duration of `45` minutes. Moreover, the cause of the restart on all the Backup pods were due to `node` issue, not the backup container.

_**Pros**_

> As the Backup pods are being scheduled independently of the etcd pods, any error within the Backup pods does not influence cluster Openshift cluster availability.

Reviewer 13: question: would a failure to create a backup block upgrades?

> Moreover, as the Backup pods are being scheduled within every master nodes, the backups are being taken from every local etcd member on the same node. This avoids overwhelming a specific member with backup requests.

Reviewer 13: question: it has been a while but doesn't a snapshot request always go to the leader?

- How to handle an empty `schedule` ?
  - This has been fixed, by using a default schedule to run the backup once at midnight.
  - This configuration could be overridden by user supplied schedule.
- How to mitigate losing backups upon control-plane scaling using `CPMSO` ?
  - This has been fixed by using a label to identify the nodes with `voting` etcd member, and using the label during scheduling the backup pods to master nodes.
- What is the minimum disk size needed in order to operate this feature without running out of disk space ?

### Implementation Details/Notes/Constraints

The _no-config_ backups builds atop of the existing [automated backup of etcd](https://docs.openshift.com/container-platform/4.15/backup_and_restore/control_plane_backup_and_restore/backing-up-etcd.html#creating-automated-etcd-backups_backup-etcd).

The `PeriodicBackupController` within **cluster-etcd-operator** checks for `Backup` CR with `name=default` within every sync. Upon applying a `Backup` CR with `name=default`, the _no-config_ backup sidecar container is being enabled within each etcd pod in Openshift cluster.

> The `Backup` configurations such as `schedule`, `timezone` and `retention policy`, are being supplied to the _no-config_ container as flags. 

Reviewer 6: So how does retention actually work for the hostpath based backup in the DS approach?

Since StatefulSet manages its own Pods, it is not possible to schedule backups by a `CronJob`. However, it is possible to generate event by the `CronJob` which is being watched by the StatefulSet. Then the backups are being taken and controlled mainly by it.

### Parallel Jobs approach

The `CronJob` based implementation schedules a `Job` upon each trigger to take the backup.
An attempt has been investigated, which exploits [Parallel Jobs](https://kubernetes.io/docs/concepts/workloads/controllers/job/#parallel-jobs).
Using NodeAffinity and PodAntiAffinity, while setting the Parallelism to the number of master nodes within an Openshift cluster, it is guaranteed that a Backup Job is being scheduled per each master node, while no more than one Backup Job is being scheduled to the same master node.
Since the `config.openshift.io/v1alpha1 Backup` API expects only one `PVC`, only one Backup Job could mount the PVC and schedule the backup.
A workaround has been tried, in which the `job.Spec` mounts a `hostPath` volume, so that it is being used on each master node independently.

> However, since the `Backup` job runs two containers (i.e. request-backup and prune) and the containers expect `PVC` as an argument, the approach has been declared as non-viable.

Reviewer 6: I don't think this is accurate based on the latest testing and discussion is it?
